<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="fitdistrplus">
<title>Overview of the fitdistrplus package • fitdistrplus</title>
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Overview of the fitdistrplus package">
<meta property="og:description" content="fitdistrplus">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">fitdistrplus</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">1.1-11</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/fitdistrplus_vignette.html">Overview of the fitdistrplus package</a>
    <a class="dropdown-item" href="../articles/Optimalgo.html">Which optimization algorithm to choose?</a>
    <a class="dropdown-item" href="../articles/FAQ.html">Frequently Asked Questions</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/aursiber/fitdistrplus/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">


<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="" class="logo" alt=""><h1>Overview of the fitdistrplus package</h1>
                        <h4 data-toc-skip class="author">Marie Laure Delignette Muller, Christophe Dutang</h4>
            
            <h4 data-toc-skip class="date">2023-05-19</h4>
      
      <small class="dont-index">Source: <a href="https://github.com/aursiber/fitdistrplus/blob/HEAD/vignettes/fitdistrplus_vignette.Rmd" class="external-link"><code>vignettes/fitdistrplus_vignette.Rmd</code></a></small>
      <div class="d-none name"><code>fitdistrplus_vignette.Rmd</code></div>
    </div>

    
    
<p><br></p>
<p>Based on the article <em>fitdistrplus: an R Package for Fitting Distributions</em>
(Marie Laure Delignette-Muller and Christophe Dutang, 2015, Journal of Statistical Software, DOI 10.18637/jss.v064.i04)</p>
<p><br></p>
<p><em><strong>Keywords</strong>:
probability distribution fitting, bootstrap, censored data, maximum likelihood, moment matching,
quantile matching, maximum goodness-of-fit, distributions, R</em></p>
<p><br></p>
<div class="section level2" number="1">
<h2 id="Introduction">
<span class="header-section-number">1</span> Introduction<a class="anchor" aria-label="anchor" href="#Introduction"></a>
</h2>
<p>Fitting distributions to data is a very common task in statistics and consists in choosing a probability distribution modelling the random variable, as well as finding parameter estimates for that distribution.
This requires judgment and expertise and generally needs an iterative process of distribution choice, parameter estimation, and quality of fit assessment.
In the R <span class="citation">(<a href="#ref-R13" role="doc-biblioref">R Development Core Team 2013</a>)</span> package <strong>MASS</strong> <span class="citation">(<a href="#ref-MASS" role="doc-biblioref">Venables and Ripley 2010</a>)</span>, maximum likelihood estimation is available via the <code>fitdistr</code> function; other steps of the fitting process can be done using other R functions <span class="citation">(<a href="#ref-Ricci05" role="doc-biblioref">Ricci 2005</a>)</span>.
In this paper, we present the R package <strong>fitdistrplus</strong> <span class="citation">(<a href="#ref-fitdistrplus" role="doc-biblioref">Delignette-Muller et al. 2014</a>)</span> implementing several methods for fitting univariate parametric distribution.
A first objective in developing this package was to provide R users a set of functions dedicated to help this overall process.</p>
<p>The <code>fitdistr</code> function estimates distribution parameters by maximizing the likelihood function using the <code>optim</code> function.
No distinction between parameters with different roles (e.g., main parameter and nuisance parameter) is made, as this paper focuses on parameter estimation from a general point-of-view.
In some cases, other estimation methods could be prefered, such as maximum goodness-of-fit estimation (also called minimum distance estimation), as proposed in the R package <strong>actuar</strong> with three different goodness-of-fit distances <span class="citation">(<a href="#ref-actuarJSS" role="doc-biblioref">Dutang, Goulet, and Pigeon 2008</a>)</span>.
While developping the <strong>fitdistrplus</strong> package, a second objective was to consider various estimation methods in addition to maximum likelihood estimation (MLE).
Functions were developped to enable moment matching estimation (MME), quantile matching estimation (QME), and maximum goodness-of-fit estimation (MGE) using eight different distances.
Moreover, the <strong>fitdistrplus</strong> package offers the possibility to specify a user-supplied function for optimization, useful in cases where classical optimization techniques, not included in <code>optim</code>, are more adequate.</p>
<p>In applied statistics, it is frequent to have to fit distributions to censored data <span class="citation">Commeau et al. (<a href="#ref-commeauetal12" role="doc-biblioref">2012</a>)</span>.
The <strong>MASS</strong> <code>fitdistr</code> function does not enable maximum likelihood estimation with this type of data.
Some packages can be used to work with censored data, especially survival data <span class="citation">Jordan (<a href="#ref-jordan05" role="doc-biblioref">2005</a>)</span>, but those packages generally focus on specific models, enabling the fit of a restricted set of distributions.
A third objective is thus to provide R users a function to estimate univariate distribution parameters from right-, left- and interval-censored data.</p>
<p>Few packages on CRAN provide estimation procedures for any user-supplied parametric distribution and support different types of data.
The <strong>distrMod</strong> package <span class="citation">(<a href="#ref-distrModJSS" role="doc-biblioref">Kohl and Ruckdeschel 2010</a>)</span> provides an object-oriented (S4) implementation of probability models and includes distribution fitting procedures for a given minimization criterion.
This criterion is a user-supplied function which is sufficiently flexible to handle censored data, yet not in a trivial way, see Example M4 of the <strong>distrMod</strong> vignette.
The fitting functions <code>MLEstimator</code> and <code>MDEstimator</code> return an S4 class for which a coercion method to class <code>mle</code> is provided so that the respective functionalities (e.g., <code>confint</code> and <code>logLik</code>) from package <strong>stats4</strong> are available, too.
In <strong>fitdistrplus</strong>, we chose to use the standard S3 class system for its understanding by most R users.
When designing the <strong>fitdistrplus</strong> package, we did not forget to implement generic functions also available for S3 classes.
Finally, various other packages provide functions to estimate the mode, the moments or the L-moments of a distribution, see the reference manuals of <strong>modeest</strong>, <strong>lmomco</strong> and <strong>Lmoments</strong> packages.</p>
<p>The package is available from the Comprehensive R Archive Network at .
The paper is organized as follows:
Section <a href="#fitnoncenscont">2</a> presents tools for fitting continuous distributions to classic non-censored data.
Section <a href="#advtopic">3</a> deals with other estimation methods and other types of data, before
Section <a href="#ccl">4</a> concludes.</p>
<p><br></p>
</div>
<div class="section level2" number="2">
<h2 id="fitnoncenscont">
<span class="header-section-number">2</span> Fitting distributions to continuous non-censored data<a class="anchor" aria-label="anchor" href="#fitnoncenscont"></a>
</h2>
<div class="section level3" number="2.1">
<h3 id="Choice">
<span class="header-section-number">2.1</span> Choice of candidate distributions<a class="anchor" aria-label="anchor" href="#Choice"></a>
</h3>
<p>For illustrating the use of various functions of the <strong>fitdistrplus</strong> package with continuous non-censored data, we will first use a data set named <code>groundbeef</code> which is included in our package.
This data set contains pointwise values of serving sizes in grams, collected in a French survey, for ground beef patties consumed by children under 5 years old.
It was used in a quantitative risk assessment published by <span class="citation">Delignette-Muller and Cornu (<a href="#ref-Delignette08" role="doc-biblioref">2008</a>)</span>.</p>
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="st"><a href="https://lbbe.univ-lyon1.fr/fr/fitdistrplus" class="external-link">"fitdistrplus"</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: MASS</span></span></code></pre>
<pre><code><span><span class="co">## Loading required package: survival</span></span></code></pre>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"groundbeef"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    254 obs. of  1 variable:</span></span>
<span><span class="co">##  $ serving: num  30 10 20 24 20 24 40 20 50 30 ...</span></span></code></pre>
<p>Before fitting one or more distributions to a data set, it is generally necessary to choose good candidates among a predefined set of distributions.
This choice may be guided by the knowledge of stochastic processes governing the modeled variable, or, in the absence of knowledge regarding the underlying process, by the observation of its empirical distribution.
To help the user in this choice, we developed functions to plot and characterize the empirical distribution.</p>
<p>First of all, it is common to start with plots of the empirical distribution function and the histogram (or density plot), which can be obtained with the <code>plotdist</code> function of the <strong>fitdistrplus</strong> package.
This function provides two plots (see Figure <a href="#fig:figgroundbeef">2.1</a>): the left-hand plot is by default the histogram on a density scale (or density plot of both, according to values of arguments <code>histo</code> and <code>demp</code>) and the right-hand plot the empirical cumulative distribution function (CDF).</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotdist.html">plotdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, histo <span class="op">=</span> <span class="cn">TRUE</span>, demp <span class="op">=</span> <span class="cn">TRUE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:figgroundbeef"></span>
<img src="fitdistrplus_vignette_files/figure-html/figgroundbeef-1.png" alt="Histogram and CDF plots of an empirical distribution for a continuous variable (serving size from the `groundbeef` data set) as provided by the `plotdist` function."><p class="caption">
Figure 2.1: Histogram and CDF plots of an empirical distribution for a continuous variable (serving size from the <code>groundbeef</code> data set) as provided by the <code>plotdist</code> function.
</p>
</div>
<p><br></p>
<p>In addition to empirical plots, descriptive statistics may help to choose candidates to describe a distribution among a set of parametric distributions.
Especially the skewness and kurtosis, linked to the third and fourth moments, are useful for this purpose.
A non-zero skewness reveals a lack of symmetry of the empirical distribution, while the kurtosis value quantifies the weight of tails in comparison to the normal distribution for which the kurtosis equals 3.
The skewness and kurtosis and their corresponding unbiased estimator <span class="citation">(<a href="#ref-casellaberger02" role="doc-biblioref">Casella and Berger 2002</a>)</span> from a sample <span class="math inline">\((X_i)_i \stackrel{\text{i.i.d.}}{\sim} X\)</span> with observations <span class="math inline">\((x_i)_i\)</span> are given by:</p>
<p><span class="math display" id="eq:eq1">\[\begin{equation}
sk(X) = \frac{E[(X-E(X))^3]}{Var(X)^{\frac{3}{2}}}~,~\widehat{sk}=\frac{\sqrt{n(n-1)}}{n-2}\times\frac{m_{3}}{m_{2}^{\frac{3}{2}}},\tag{2.1}
\end{equation}\]</span></p>
<p><span class="math display" id="eq:eq2">\[\begin{equation}
kr(X) = \frac{E[(X-E(X))^4]}{Var(X)^{2}}~,~\widehat{kr}=\frac{n-1}{(n-2)(n-3)}((n+1) \times \frac{m_{4}}{m_{2}^{2}}-3(n-1)) + 3,\tag{2.2}
\end{equation}\]</span></p>
<p>where <span class="math inline">\(m_{2}\)</span>, <span class="math inline">\(m_{3}\)</span>, <span class="math inline">\(m_{4}\)</span> denote empirical moments defined by <span class="math inline">\(m_{k}=\frac{1}{n}\sum_{i=1}^n(x_{i}-\overline{x})^{k}\)</span>, with <span class="math inline">\(x_{i}\)</span> the <span class="math inline">\(n\)</span> observations of variable <span class="math inline">\(x\)</span> and <span class="math inline">\(\overline{x}\)</span> their mean value.</p>
<p>The <code>descdist</code> function provides classical descriptive statistics (minimum, maximum, median, mean, standard deviation), skewness and kurtosis.
By default, unbiased estimations of the three last statistics are provided.
Nevertheless, the argument <code>method</code> can be changed from <code>"unbiased"</code> (default) to <code>"sample"</code> to obtain them without correction for bias.
A skewness-kurtosis plot such as the one proposed by <span class="citation">Cullen and Frey (<a href="#ref-Cullen99" role="doc-biblioref">1999</a>)</span> is provided by the <code>descdist</code> function for the empirical distribution (see Figure <a href="#fig:descgroundbeefplot">2.2</a> for the <code>groundbeef</code> data set).
On this plot, values for common distributions are displayed in order to help the choice of distributions to fit to data.
For some distributions (normal, uniform, logistic, exponential), there is only one possible value for the skewness and the kurtosis.
Thus, the distribution is represented by a single point on the plot. For other distributions, areas of possible values are represented, consisting in lines (as for gamma and lognormal distributions), or larger areas (as for beta distribution).</p>
<p>Skewness and kurtosis are known not to be robust.
In order to take into account the uncertainty of the estimated values of kurtosis and skewness from data, a nonparametric bootstrap procedure <span class="citation">(<a href="#ref-efrontibshirani94" role="doc-biblioref">Efron and Tibshirani 1994</a>)</span> can be performed by using the argument <code>boot</code>.
Values of skewness and kurtosis are computed on bootstrap samples (constructed by random sampling with replacement from the original data set) and reported on the skewness-kurtosis plot.
Nevertheless, the user needs to know that skewness and kurtosis, like all higher moments, have a very high variance.
This is a problem which cannot be completely solved by the use of bootstrap.
The skewness-kurtosis plot should then be regarded as indicative only.
The properties of the random variable should be considered, notably its expected value and its range, as a complement to the use of the <code>plotdist</code> and <code>descdist</code> functions.
Below is a call to the <code>descdist</code> function to describe the distribution of the serving size from the <code>groundbeef</code> data set and to draw the corresponding skewness-kurtosis plot (see Figure <a href="#fig:descgroundbeefplot">2.2</a>).
Looking at the results on this example with a positive skewness and a kurtosis not far from 3, the fit of three common right-skewed distributions could be considered, Weibull, gamma and lognormal distributions.</p>
<div class="sourceCode" id="cb7"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/descdist.html">descdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, boot <span class="op">=</span> <span class="fl">1000</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:descgroundbeefplot"></span>
<img src="fitdistrplus_vignette_files/figure-html/descgroundbeefplot-1.png" alt="Skewness-kurtosis plot for a continuous variable (serving size from the `groundbeef` data set) as provided by the `descdist` function."><p class="caption">
Figure 2.2: Skewness-kurtosis plot for a continuous variable (serving size from the <code>groundbeef</code> data set) as provided by the <code>descdist</code> function.
</p>
</div>
<pre><code><span><span class="co">## summary statistics</span></span>
<span><span class="co">## ------</span></span>
<span><span class="co">## min:  10   max:  200 </span></span>
<span><span class="co">## median:  79 </span></span>
<span><span class="co">## mean:  73.65 </span></span>
<span><span class="co">## estimated sd:  35.88 </span></span>
<span><span class="co">## estimated skewness:  0.7353 </span></span>
<span><span class="co">## estimated kurtosis:  3.551</span></span></code></pre>
</div>
<div class="section level3" number="2.2">
<h3 id="FIT">
<span class="header-section-number">2.2</span> Fit of distributions by maximum likelihood estimation<a class="anchor" aria-label="anchor" href="#FIT"></a>
</h3>
<p>Once selected, one or more parametric distributions <span class="math inline">\(f(.\vert \theta)\)</span> (with parameter <span class="math inline">\(\theta\in\mathbb{R}^d\)</span>) may be fitted to the data set, one at a time, using the <code>fitdist</code> function.
Under the i.i.d. sample assumption, distribution parameters <span class="math inline">\(\theta\)</span> are by default estimated by maximizing the likelihood function defined as:</p>
<p><span class="math display" id="eq:eq3">\[\begin{equation}
L(\theta)=\prod_{i=1}^n f(x_{i}\vert \theta)\tag{2.3}
\end{equation}\]</span></p>
<p>with <span class="math inline">\(x_{i}\)</span> the <span class="math inline">\(n\)</span> observations of variable <span class="math inline">\(X\)</span> and <span class="math inline">\(f(.\vert \theta)\)</span> the density function of the parametric distribution.
The other proposed estimation methods are described in Section <a href="#Alternatives">3.1</a>.</p>
<p>The <code>fitdist</code> function returns an S3 object of class <code>fitdist</code> for which <code>print</code>, <code>summary</code> and <code>plot</code> functions are provided.
The fit of a distribution using <code>fitdist</code> assumes that the corresponding <code>d</code>, <code>p</code>, <code>q</code> functions (standing respectively for the density, the distribution and the quantile functions) are defined.
Classical distributions are already defined in that way in the <strong>stats</strong> package, e.g., <code>dnorm</code>, <code>pnorm</code> and <code>qnorm</code> for the normal distribution (see <code><a href="https://rdrr.io/r/stats/Distributions.html" class="external-link">?Distributions</a></code>).
Others may be found in various packages (see the CRAN task view: Probability Distributions at ).
Distributions not found in any package must be implemented by the user as <code>d</code>, <code>p</code>, <code>q</code> functions.
In the call to <code>fitdist</code>, a distribution has to be specified via the argument <code>dist</code> either by the character string corresponding to its common root name used in the names of <code>d</code>, <code>p</code>, <code>q</code> functions (e.g., <code>"norm"</code> for the normal distribution) or by the density function itself, from which the root name is extracted (e.g., <code>dnorm</code> for the normal distribution).
Numerical results returned by the <code>fitdist</code> function are
(1) the parameter estimates,
(2) the estimated standard errors (computed from the estimate of the Hessian matrix at the maximum likelihood solution),
(3) the loglikelihood,
(4) Akaike and Bayesian information criteria (the so-called AIC and BIC), and (5) the correlation matrix between parameter estimates.
Below is a call to the <code>fitdist</code> function to fit a Weibull distribution to the serving size from the <code>groundbeef</code> data set.</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fw</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"weibull"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fw</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fitting of the distribution ' weibull ' by maximum likelihood </span></span>
<span><span class="co">## Parameters : </span></span>
<span><span class="co">##       estimate Std. Error</span></span>
<span><span class="co">## shape    2.186     0.1046</span></span>
<span><span class="co">## scale   83.348     2.5269</span></span>
<span><span class="co">## Loglikelihood:  -1255   AIC:  2514   BIC:  2522 </span></span>
<span><span class="co">## Correlation matrix:</span></span>
<span><span class="co">##        shape  scale</span></span>
<span><span class="co">## shape 1.0000 0.3218</span></span>
<span><span class="co">## scale 0.3218 1.0000</span></span></code></pre>
<p>The plot of an object of class <code>fitdist</code> provides four classical goodness-of-fit plots <span class="citation">(<a href="#ref-Cullen99" role="doc-biblioref">Cullen and Frey 1999</a>)</span> presented on Figure <a href="#fig:groundbeefcomp">2.3</a>:</p>
<ul>
<li>a density plot representing the density function of the fitted distribution along with the histogram of the empirical distribution,</li>
<li>a CDF plot of both the empirical distribution and the fitted distribution,</li>
<li>a Q-Q plot representing the empirical quantiles (y-axis) against the theoretical quantiles (x-axis),</li>
<li>a P-P plot representing the empirical distribution function evaluated at each data point (y-axis) against the fitted distribution function (x-axis).</li>
</ul>
<p>For CDF, Q-Q and P-P plots, the probability plotting position is defined by default using Hazen’s rule, with probability points of the empirical distribution calculated as <code>(1:n - 0.5)/n</code>, as recommended by <span class="citation">Blom (<a href="#ref-Blom" role="doc-biblioref">1959</a>)</span>.
This plotting position can be easily changed (see the reference manual for details <span class="citation">(<a href="#ref-fitdistrplus" role="doc-biblioref">Delignette-Muller et al. 2014</a>)</span>).</p>
<p>Unlike the generic <code>plot</code> function, the <code>denscomp</code>, <code>cdfcomp</code>, <code>qqcomp</code> and <code>ppcomp</code> functions enable to draw separately each of these four plots, in order to compare the empirical distribution and multiple parametric distributions fitted on a same data set.
These functions must be called with a first argument corresponding to a list of objects of class <code>fitdist</code>, and optionally further arguments to customize the plot (see the reference manual for lists of arguments that may be specific to each plot <span class="citation">(<a href="#ref-fitdistrplus" role="doc-biblioref">Delignette-Muller et al. 2014</a>)</span>).
In the following example, we compare the fit of a Weibull, a lognormal and a gamma distributions to the <code>groundbeef</code> data set (Figure <a href="#fig:groundbeefcomp">2.3</a>).</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span>, mar <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">4</span>, <span class="fl">4</span>, <span class="fl">2</span>, <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fg</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="va">fln</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"lnorm"</span><span class="op">)</span></span>
<span><span class="va">plot.legend</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Weibull"</span>, <span class="st">"lognormal"</span>, <span class="st">"gamma"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">denscomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fw</span>, <span class="va">fln</span>, <span class="va">fg</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="va">plot.legend</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">qqcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fw</span>, <span class="va">fln</span>, <span class="va">fg</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="va">plot.legend</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fw</span>, <span class="va">fln</span>, <span class="va">fg</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="va">plot.legend</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">ppcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fw</span>, <span class="va">fln</span>, <span class="va">fg</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="va">plot.legend</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:groundbeefcomp"></span>
<img src="fitdistrplus_vignette_files/figure-html/groundbeefcomp-1.png" alt="Four Goodness-of-fit plots for various distributions fitted to continuous data (Weibull, gamma and lognormal distributions fitted to serving sizes from the `groundbeef` data set) as provided by functions `denscomp`, `qqcomp`, `cdfcomp` and `ppcomp`."><p class="caption">
Figure 2.3: Four Goodness-of-fit plots for various distributions fitted to continuous data (Weibull, gamma and lognormal distributions fitted to serving sizes from the <code>groundbeef</code> data set) as provided by functions <code>denscomp</code>, <code>qqcomp</code>, <code>cdfcomp</code> and <code>ppcomp</code>.
</p>
</div>
<p><br></p>
<p>The density plot and the CDF plot may be considered as the basic classical goodness-of-fit plots.
The two other plots are complementary and can be very informative in some cases.
The Q-Q plot emphasizes the lack-of-fit at the distribution tails while the P-P plot emphasizes the lack-of-fit at the distribution center.
In the present example (in Figure <a href="#fig:groundbeefcomp">2.3</a>), none of the three fitted distributions correctly describes the center of the distribution, but the Weibull and gamma distributions
could be prefered for their better description of the right tail of the empirical distribution, especially if this tail is important in the use of the fitted distribution, as it is in the context of food risk assessment.</p>
<p>The data set named <code>endosulfan</code> will now be used to illustrate other features of the <strong>fitdistrplus</strong> package.
This data set contains acute toxicity values for the organochlorine pesticide endosulfan (geometric mean of LC50 ou EC50 values in <span class="math inline">\(\mu g.L^{-1}\)</span>), tested on Australian and non-Australian laboratory-species <span class="citation">(<a href="#ref-Hose04" role="doc-biblioref">Hose and Van den Brink 2004</a>)</span>.
In ecotoxicology, a lognormal or a loglogistic distribution is often fitted to such a data set in order to characterize the species sensitivity distribution (SSD) for a pollutant.
A low percentile of the fitted distribution, generally the 5% percentile, is then calculated and named the hazardous concentration 5% (HC5).
It is interpreted as the value of the pollutant concentration protecting 95% of the species <span class="citation">(<a href="#ref-Posthuma2010" role="doc-biblioref">Posthuma, Suter, and Traas 2010</a>)</span>.
But the fit of a lognormal or a loglogistic distribution to the whole <code>endosulfan</code> data set is rather bad (Figure <a href="#fig:fitendo">2.4</a>), especially due to a minority of very high values.
The two-parameter Pareto distribution and the three-parameter Burr distribution (which is an extension of both the loglogistic and the Pareto distributions) have been fitted.
Pareto and Burr distributions are provided in the package <strong>actuar</strong>.
Until here, we did not have to define starting values (in the optimization process) as reasonable starting values are implicity defined within the <code>fitdist</code> function for most of the distributions defined in R (see <code><a href="../reference/fitdist.html">?fitdist</a></code> for details).
For other distributions like the Pareto and the Burr distribution, initial values for the distribution parameters have to be supplied in the argument <code>start</code>, as a named list with initial values for each parameter (as they appear in the <code>d</code>, <code>p</code>, <code>q</code> functions).
Having defined reasonable starting values<a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content='&lt;p&gt;The &lt;code&gt;plotdist&lt;/code&gt; function can plot any parametric distribution with specified parameter values in argument &lt;code&gt;para&lt;/code&gt;. It can thus help to find correct initial values for the distribution parameters in non trivial cases, by iterative calls if necessary (see the reference
manual for examples &lt;span class="citation"&gt;(&lt;a href="#ref-fitdistrplus" role="doc-biblioref"&gt;Delignette-Muller et al. 2014&lt;/a&gt;)&lt;/span&gt;).&lt;/p&gt;'><sup>1</sup></a> various distributions can be fitted and graphically compared.
On this example, the function <code>cdfcomp</code> can be used to report CDF values in a logscale so as to emphasize discrepancies on the tail of interest while defining an HC5 value (Figure <a href="#fig:fitendo">2.4</a>).</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.com/vigou3/actuar" class="external-link">actuar</a></span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## </span></span>
<span><span class="co">## Attaching package: 'actuar'</span></span></code></pre>
<pre><code><span><span class="co">## The following objects are masked from 'package:stats':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     sd, var</span></span></code></pre>
<pre><code><span><span class="co">## The following object is masked from 'package:grDevices':</span></span>
<span><span class="co">## </span></span>
<span><span class="co">##     cm</span></span></code></pre>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"endosulfan"</span><span class="op">)</span></span>
<span><span class="va">ATV</span> <span class="op">&lt;-</span> <span class="va">endosulfan</span><span class="op">$</span><span class="va">ATV</span></span>
<span><span class="va">fendo.ln</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"lnorm"</span><span class="op">)</span></span>
<span><span class="va">fendo.ll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"llogis"</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, scale <span class="op">=</span> <span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fendo.P</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"pareto"</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">1</span>, scale <span class="op">=</span> <span class="fl">500</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fendo.B</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"burr"</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape1 <span class="op">=</span> <span class="fl">0.3</span>, shape2 <span class="op">=</span> <span class="fl">1</span>, rate <span class="op">=</span> <span class="fl">1</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fendo.ln</span>, <span class="va">fendo.ll</span>, <span class="va">fendo.P</span>, <span class="va">fendo.B</span><span class="op">)</span>, xlogscale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>        ylogscale <span class="op">=</span> <span class="cn">TRUE</span>, legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lognormal"</span>, <span class="st">"loglogistic"</span>, <span class="st">"Pareto"</span>, <span class="st">"Burr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:fitendo"></span>
<img src="fitdistrplus_vignette_files/figure-html/fitendo-1.png" alt="CDF plot to compare the fit of four distributions to acute toxicity values of various organisms for the organochlorine pesticide endosulfan (`endosulfan` data set) as provided by the `cdfcomp` function, with CDF values in a logscale to emphasize discrepancies on the left tail."><p class="caption">
Figure 2.4: CDF plot to compare the fit of four distributions to acute toxicity values of various organisms for the organochlorine pesticide endosulfan (<code>endosulfan</code> data set) as provided by the <code>cdfcomp</code> function, with CDF values in a logscale to emphasize discrepancies on the left tail.
</p>
</div>
<p><br></p>
<p>None of the fitted distribution correctly describes the right tail observed in the data set, but as shown in Figure <a href="#fig:fitendo">2.4</a>, the left-tail seems to be better described by the Burr distribution.
Its use could then be considered to estimate the HC5 value as the 5% quantile of the distribution.
This can be easily done using the <code>quantile</code> generic function defined for an object of class <code>fitdist</code>.
Below is this calculation together with the calculation of the empirical quantile for comparison.</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">fendo.B</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Estimated quantiles for each specified probability (non-censored data)</span></span>
<span><span class="co">##          p=0.05</span></span>
<span><span class="co">## estimate 0.2939</span></span></code></pre>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">ATV</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##  5% </span></span>
<span><span class="co">## 0.2</span></span></code></pre>
<p>In addition to the ecotoxicology context, the <code>quantile</code> generic function is also attractive in the actuarial-financial context.
In fact, the value-at-risk <span class="math inline">\(VAR_\alpha\)</span> is defined as the <span class="math inline">\(1-\alpha\)</span>-quantile of the loss distribution and can be computed with <code>quantile</code> on a <code>fitdist</code> object.</p>
<p>The computation of different goodness-of-fit statistics is proposed in the <strong>fitdistrplus</strong> package in order to further compare fitted distributions.
The purpose of goodness-of-fit statistics aims to measure the distance between the fitted parametric distribution and the empirical distribution: e.g., the distance between the fitted cumulative distribution function <span class="math inline">\(F\)</span> and the empirical distribution function <span class="math inline">\(F_{n}\)</span>.
When fitting continuous distributions, three goodness-of-fit statistics are classicaly considered: Cramer-von Mises, Kolmogorov-Smirnov and Anderson-Darling statistics <span class="citation">(<a href="#ref-Stephens86" role="doc-biblioref">D’Agostino and Stephens 1986</a>)</span>.
Naming <span class="math inline">\(x_{i}\)</span> the <span class="math inline">\(n\)</span> observations of a continuous variable <span class="math inline">\(X\)</span> arranged in an ascending order, Table <a href="#tab:tabKSCvMAD">2.1</a> gives the definition and the empirical estimate of the three considered goodness-of-fit statistics.
They can be computed using the function <code>gofstat</code> as defined by Stephens <span class="citation">(<a href="#ref-Stephens86" role="doc-biblioref">D’Agostino and Stephens 1986</a>)</span>.</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gofstat.html">gofstat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fendo.ln</span>, <span class="va">fendo.ll</span>, <span class="va">fendo.P</span>, <span class="va">fendo.B</span><span class="op">)</span>, </span>
<span>        fitnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lnorm"</span>, <span class="st">"llogis"</span>, <span class="st">"Pareto"</span>, <span class="st">"Burr"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Goodness-of-fit statistics</span></span>
<span><span class="co">##                               lnorm llogis  Pareto    Burr</span></span>
<span><span class="co">## Kolmogorov-Smirnov statistic 0.1672 0.1196 0.08488 0.06155</span></span>
<span><span class="co">## Cramer-von Mises statistic   0.6374 0.3827 0.13926 0.06803</span></span>
<span><span class="co">## Anderson-Darling statistic   3.4721 2.8316 0.89206 0.52393</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Goodness-of-fit criteria</span></span>
<span><span class="co">##                                lnorm llogis Pareto Burr</span></span>
<span><span class="co">## Akaike's Information Criterion  1069   1069   1048 1046</span></span>
<span><span class="co">## Bayesian Information Criterion  1074   1075   1053 1054</span></span></code></pre>
<p><br></p>
<table class="table">
<caption>
<span id="tab:tabKSCvMAD">Table 2.1: </span> Goodness-of-fit statistics as defined by Stephens <span class="citation">(<a href="#ref-Stephens86" role="doc-biblioref">D’Agostino and Stephens 1986</a>)</span>.</caption>
<colgroup>
<col width="9%">
<col width="29%">
<col width="60%">
</colgroup>
<thead><tr class="header">
<th>Statistic</th>
<th>General formula</th>
<th>Computational formula</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Kolmogorov-Smirnov (KS)</td>
<td><span class="math inline">\(\sup|F_{n}(x) - F(x)|\)</span></td>
<td>
<span class="math inline">\(\max(D^{+},D^{-})\)</span> with <span class="math inline">\(D^{+}=\max\limits_{i=1,\dots,n}\left(\frac{i}{n} - F_i\right)\)</span> and <span class="math inline">\(D^{-}=\max\limits_{i=1,\dots,n}\left(F_{i}-\frac{i-1}{n}\right)\)</span>
</td>
</tr>
<tr class="even">
<td>Cramer-von Mises (CvM)</td>
<td><span class="math inline">\(n \int_{-\infty}^{\infty}(F_{n}(x) - F(x))^2 dx\)</span></td>
<td><span class="math inline">\(\frac{1}{12n} + \sum\limits_{i=1}^n \left(F_i-\frac{2i-1}{2n} \right)^{2}\)</span></td>
</tr>
<tr class="odd">
<td>Anderson-Darling (AD)</td>
<td><span class="math inline">\(n \int_{-\infty}^{\infty}\frac{(F_{n}(x) - F(x))^2}{F(x) (1 - F(x))} dx\)</span></td>
<td><span class="math inline">\(-n -\frac{1}{n}\sum\limits_{i=1}^n (2i-1)\log(F_i(1-F_{n+1-i}))\)</span></td>
</tr>
</tbody>
</table>
<p>where <span class="math inline">\(F_i\stackrel{\triangle}{=} F(x_i)\)</span></p>
<p><br></p>
<p>As giving more weight to distribution tails, the Anderson-Darling statistic is of special interest when it matters to equally emphasize the tails as well as the main body of a distribution.
This is often the case in risk assessment <span class="citation">Vose (<a href="#ref-Vose10" role="doc-biblioref">2010</a>)</span>.
For this reason, this statistics is often used to select the best distribution among those fitted.
Nevertheless, this statistics should be used cautiously when comparing fits of various distributions.
Keeping in mind that the weighting of each CDF quadratic difference depends on the parametric distribution in its definition (see Table <a href="#tab:tabKSCvMAD">2.1</a>), Anderson-Darling statistics computed for several distributions fitted on a same data set are theoretically difficult to compare.
Moreover, such a statistic, as Cramer-von Mises and Kolmogorov-Smirnov ones, does not take into account the complexity of the model (i.e., parameter number).
It is not a problem when compared distributions are characterized by the same number of parameters, but it could systematically promote the selection of the more complex distributions in the other case.
Looking at classical penalized criteria based on the loglikehood (AIC, BIC) seems thus also interesting, especially to discourage overfitting.</p>
<p>In the previous example, all the goodness-of-fit statistics based on the CDF distance are in favor of the Burr distribution, the only one characterized by three parameters, while AIC and BIC values respectively give the preference to the Burr distribution or the Pareto distribution.
The choice between these two distributions seems thus less obvious and could be discussed.
Even if specifically recommended for discrete distributions, the Chi-squared statistic may also be used for continuous distributions (see Section <a href="#otherdata">3.3</a> and the reference manual for examples <span class="citation">(<a href="#ref-fitdistrplus" role="doc-biblioref">Delignette-Muller et al. 2014</a>)</span>).</p>
</div>
<div class="section level3" number="2.3">
<h3 id="Uncertainty">
<span class="header-section-number">2.3</span> Uncertainty in parameter estimates<a class="anchor" aria-label="anchor" href="#Uncertainty"></a>
</h3>
<p>The uncertainty in the parameters of the fitted distribution can be estimated by parametric or nonparametric bootstraps using the <code>boodist</code> function for non-censored data <span class="citation">(<a href="#ref-efrontibshirani94" role="doc-biblioref">Efron and Tibshirani 1994</a>)</span>.
This function returns the bootstrapped values of parameters in an S3 class object which can be plotted to visualize the bootstrap region.
The medians and the 95% confidence intervals of parameters (2.5 and 97.5 percentiles) are printed in the summary.
When inferior to the whole number of iterations (due to lack of convergence of the optimization algorithm for some bootstrapped data sets), the number of iterations for which the estimation converges is also printed in the summary.</p>
<p>The plot of an object of class <code>bootdist</code> consists in a scatterplot or a matrix of scatterplots of the bootstrapped values of parameters providing a representation of the joint uncertainty distribution of the fitted parameters.
Below is an example of the use of the <code>bootdist</code> function with the previous fit of the Burr distribution to the <code>endosulfan</code> data set (Figure <a href="#fig:bootstrap">2.5</a>).</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bendo.B</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/bootdist.html">bootdist</a></span><span class="op">(</span><span class="va">fendo.B</span>, niter <span class="op">=</span> <span class="fl">1001</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">bendo.B</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Parametric bootstrap medians and 95% percentile CI </span></span>
<span><span class="co">##        Median    2.5%  97.5%</span></span>
<span><span class="co">## shape1 0.1983 0.09283 0.3606</span></span>
<span><span class="co">## shape2 1.5863 1.05306 3.0629</span></span>
<span><span class="co">## rate   1.4907 0.70828 2.7775</span></span></code></pre>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/plot.default.html" class="external-link">plot</a></span><span class="op">(</span><span class="va">bendo.B</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:bootstrap"></span>
<img src="fitdistrplus_vignette_files/figure-html/bootstrap-1.png" alt="Bootstrappped values of parameters for a fit of the Burr distribution characterized by three parameters (example on the `endosulfan` data set) as provided by the plot of an object of class `bootdist`."><p class="caption">
Figure 2.5: Bootstrappped values of parameters for a fit of the Burr distribution characterized by three parameters (example on the <code>endosulfan</code> data set) as provided by the plot of an object of class <code>bootdist</code>.
</p>
</div>
<p><br></p>
<p>Bootstrap samples of parameter estimates are useful especially to calculate confidence intervals on each parameter of the fitted distribution from the marginal distribution of the bootstraped values.
It is also interesting to look at the joint distribution of the bootstraped values in a scatterplot (or a matrix of scatterplots if the number of parameters exceeds two) in order to understand the potential structural correlation between parameters (see Figure <a href="#fig:bootstrap">2.5</a>).</p>
<p>The use of the whole bootstrap sample is also of interest in the risk assessment field.
Its use enables the characterization of uncertainty in distribution parameters.
It can be directly used within a second-order Monte Carlo simulation framework, especially within the package <strong>mc2d</strong> <span class="citation">(<a href="#ref-mc2d" role="doc-biblioref">Pouillot, Delignette-Muller, and Denis 2011</a>)</span>.
One could refer to <span class="citation">Pouillot and Delignette-Muller (<a href="#ref-Pouillot10" role="doc-biblioref">2010</a>)</span> for an introduction to the use of <strong>mc2d</strong> and <strong>fitdistrplus</strong> packages in the context of quantitative risk assessment.</p>
<p>The bootstrap method can also be used to calculate confidence intervals on quantiles of the fitted distribution.
For this purpose, a generic <code>quantile</code> function is provided for class <code>bootdist</code>.
By default, 95% percentiles bootstrap confidence intervals of quantiles are provided.
Going back to the previous example from ecotoxicolgy, this function can be used to estimate the uncertainty associated to the HC5 estimation, for example from the previously fitted Burr distribution to the <code>endosulfan</code> data set.</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">bendo.B</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## (original) estimated quantiles for each specified probability (non-censored data)</span></span>
<span><span class="co">##          p=0.05</span></span>
<span><span class="co">## estimate 0.2939</span></span>
<span><span class="co">## Median of bootstrap estimates</span></span>
<span><span class="co">##          p=0.05</span></span>
<span><span class="co">## estimate 0.2994</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## two-sided 95 % CI of each quantile</span></span>
<span><span class="co">##        p=0.05</span></span>
<span><span class="co">## 2.5 %  0.1792</span></span>
<span><span class="co">## 97.5 % 0.4999</span></span></code></pre>
</div>
</div>
<div class="section level2" number="3">
<h2 id="advtopic">
<span class="header-section-number">3</span> Advanced topics<a class="anchor" aria-label="anchor" href="#advtopic"></a>
</h2>
<div class="section level3" number="3.1">
<h3 id="Alternatives">
<span class="header-section-number">3.1</span> Alternative methods for parameter estimation<a class="anchor" aria-label="anchor" href="#Alternatives"></a>
</h3>
<p>This subsection focuses on alternative estimation methods.
One of the alternative for continuous distributions is the maximum goodness-of-fit estimation method also called minimum distance estimation method <span class="citation">Dutang, Goulet, and Pigeon (<a href="#ref-actuarJSS" role="doc-biblioref">2008</a>)</span>.
In this package this method is proposed with eight different distances: the three classical distances defined in Table <a href="#tab:tabKSCvMAD">2.1</a>,
or one of the variants of the Anderson-Darling distance proposed by <span class="citation">Luceno (<a href="#ref-Luceno06" role="doc-biblioref">2006</a>)</span> and defined in Table <a href="#tab:modifiedAD">3.1</a>.
The right-tail AD gives more weight to the right-tail, the left-tail AD gives more weight only to the left tail. Either of the tails, or both of them, can receive even larger weights by using second order Anderson-Darling Statistics.</p>
<p><br></p>
<table class="table">
<caption>
<span id="tab:modifiedAD">Table 3.1: </span> Modified Anderson-Darling statistics as defined by <span class="citation">Luceno (<a href="#ref-Luceno06" role="doc-biblioref">2006</a>)</span>.</caption>
<colgroup>
<col width="14%">
<col width="35%">
<col width="49%">
</colgroup>
<thead><tr class="header">
<th>Statistic</th>
<th>General formula</th>
<th>Computational formula</th>
</tr></thead>
<tbody>
<tr class="odd">
<td>Right-tail AD (ADR)</td>
<td><span class="math inline">\(\int_{-\infty}^{\infty}\frac{(F_{n}(x) - F(x))^2 }{1 - F(x)} dx\)</span></td>
<td><span class="math inline">\(\frac{n}{2} -2\sum\limits_{i=1}^nF_i -\frac{1}{n}\sum\limits_{i=1}^n(2i-1)ln(\overline F_{n+1-i})\)</span></td>
</tr>
<tr class="even">
<td>Left-tail AD (ADL)</td>
<td><span class="math inline">\(\int_{-\infty}^{\infty}\frac{(F_{n}(x) - F(x))^2 }{(F(x))} dx\)</span></td>
<td><span class="math inline">\(-\frac{3n}{2} +2\sum\limits_{i=1}^nF_i -\frac{1}{n}\sum\limits_{i=1}^n(2i-1)ln(F_i)\)</span></td>
</tr>
<tr class="odd">
<td>Right-tail AD 2nd order (AD2R)</td>
<td><span class="math inline">\(ad2r=\int_{-\infty}^{\infty}\frac{(F_{n}(x) - F(x))^2 }{(1 - F(x))^{2}} dx\)</span></td>
<td><span class="math inline">\(ad2r=2\sum\limits_{i=1}^nln(\overline F_i) +\frac{1}{n}\sum\limits_{i=1}^n \frac{2i-1}{\overline F_{n+1-i}}\)</span></td>
</tr>
<tr class="even">
<td>Left-tail AD 2nd order (AD2L)</td>
<td><span class="math inline">\(ad2l=\int_{-\infty}^{\infty}\frac{(F_{n}(x) - F(x))^2 }{(F(x))^{2}} dx\)</span></td>
<td><span class="math inline">\(ad2l=2\sum\limits_{i=1}^nln(F_i) +\frac{1}{n}\sum\limits_{i=1}^n\frac{2i-1}{F_i}\)</span></td>
</tr>
<tr class="odd">
<td>AD 2nd order (AD2)</td>
<td><span class="math inline">\(ad2r+ad2l\)</span></td>
<td><span class="math inline">\(ad2r+ad2l\)</span></td>
</tr>
</tbody>
</table>
<p>where <span class="math inline">\(F_i\stackrel{\triangle}{=} F(x_{i})\)</span> and <span class="math inline">\(\overline F_i\stackrel{\triangle}{=}1-F(x_{i})\)</span></p>
<p><br></p>
<p>To fit a distribution by maximum goodness-of-fit estimation, one needs to fix the argument <code>method</code> to <code>mge</code> in the call to <code>fitdist</code> and to specify the argument <code>gof</code> coding for the chosen goodness-of-fit distance.
This function is intended to be used only with continuous non-censored data.</p>
<p>Maximum goodness-of-fit estimation may be useful to give more weight to data at one tail of the distribution.
In the previous example from ecotoxicology, we used a non classical distribution (the Burr distribution) to correctly fit the empirical distribution especially on its left tail.
In order to correctly estimate the 5<span class="math inline">\(\%\)</span> percentile, we could also consider the fit of the classical lognormal distribution, but minimizing a goodness-of-fit distance giving more weight to the left tail of the empirical distribution.
In what follows, the left tail Anderson-Darling distances of first or second order are used to fit a lognormal to <code>endosulfan</code> data set (see Figure <a href="#fig:plotfitMGE">3.1</a>).</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fendo.ln.ADL</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"lnorm"</span>, method <span class="op">=</span> <span class="st">"mge"</span>, gof <span class="op">=</span> <span class="st">"ADL"</span><span class="op">)</span></span>
<span><span class="va">fendo.ln.AD2L</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">ATV</span>, <span class="st">"lnorm"</span>, method <span class="op">=</span> <span class="st">"mge"</span>, gof <span class="op">=</span> <span class="st">"AD2L"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fendo.ln</span>, <span class="va">fendo.ln.ADL</span>, <span class="va">fendo.ln.AD2L</span><span class="op">)</span>, </span>
<span>  xlogscale <span class="op">=</span> <span class="cn">TRUE</span>, ylogscale <span class="op">=</span> <span class="cn">TRUE</span>, </span>
<span>  main <span class="op">=</span> <span class="st">"Fitting a lognormal distribution"</span>,</span>
<span>  xlegend <span class="op">=</span> <span class="st">"bottomright"</span>, </span>
<span>  legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MLE"</span>, <span class="st">"Left-tail AD"</span>, <span class="st">"Left-tail AD 2nd order"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:plotfitMGE"></span>
<img src="fitdistrplus_vignette_files/figure-html/plotfitMGE-1.png" alt="Comparison of a lognormal distribution fitted by MLE and by MGE using two different goodness-of-fit distances: left-tail Anderson-Darling and left-tail Anderson Darling of second order (example with the `endosulfan` data set) as provided by the `cdfcomp` function, with CDF values in a logscale to emphasize discrepancies on the left tail."><p class="caption">
Figure 3.1: Comparison of a lognormal distribution fitted by MLE and by MGE using two different goodness-of-fit distances: left-tail Anderson-Darling and left-tail Anderson Darling of second order (example with the <code>endosulfan</code> data set) as provided by the <code>cdfcomp</code> function, with CDF values in a logscale to emphasize discrepancies on the left tail.
</p>
</div>
<p><br></p>
<p>Comparing the 5% percentiles (HC5) calculated using these three fits to the one calculated from the MLE fit of the Burr distribution, we can observe, on this example, that fitting the lognormal distribution by maximizing left tail Anderson-Darling distances of first or second order enables to approach the value obtained by fitting the Burr distribution by MLE.</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">HC5.estimates</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  empirical <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">ATV</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  Burr <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">fendo.B</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">quantiles</span><span class="op">)</span>, </span>
<span>  lognormal_MLE <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">fendo.ln</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">quantiles</span><span class="op">)</span>, </span>
<span>  lognormal_AD2 <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">fendo.ln.ADL</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">quantiles</span><span class="op">)</span>, </span>
<span>  lognormal_AD2L <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/numeric.html" class="external-link">as.numeric</a></span><span class="op">(</span><span class="fu"><a href="../reference/quantile.html">quantile</a></span><span class="op">(</span><span class="va">fendo.ln.AD2L</span>, probs <span class="op">=</span> <span class="fl">0.05</span><span class="op">)</span><span class="op">$</span><span class="va">quantiles</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##      empirical           Burr  lognormal_MLE  lognormal_AD2 lognormal_AD2L </span></span>
<span><span class="co">##        0.20000        0.29393        0.07259        0.19591        0.25877</span></span></code></pre>
<p>The moment matching estimation (MME) is another method commonly used to fit parametric distributions <span class="citation">(<a href="#ref-Vose10" role="doc-biblioref">Vose 2010</a>)</span>.
MME consists in finding the value of the parameter <span class="math inline">\(\theta\)</span> that equalizes the first theoretical raw moments of the parametric distribution to the corresponding empirical raw moments as in Equation <a href="#eq:eq4">(3.1)</a>:</p>
<p><span class="math display" id="eq:eq4">\[\begin{equation}
E(X^{k}|\theta)=\frac{1}{n}\sum_{i=1}^{n}x_{i}^{k},\tag{3.1}
\end{equation}\]</span>
for <span class="math inline">\(k=1,\ldots,d\)</span>, with <span class="math inline">\(d\)</span> the number of parameters to estimate and <span class="math inline">\(x_{i}\)</span> the <span class="math inline">\(n\)</span> observations of variable <span class="math inline">\(X\)</span>.
For moments of order greater than or equal to 2, it may also be relevant to match centered moments.
Therefore, we match the moments given in Equation <a href="#eq:eq5">(3.2)</a>:</p>
<p><span class="math display" id="eq:eq5">\[\begin{equation}
E(X\vert \theta) = \overline{x} ~,~E\left((X-E(X))^{k}|\theta\right)=m_k, \text{ for }  k=2,\ldots,d,\tag{3.2}
\end{equation}\]</span></p>
<p>where <span class="math inline">\(m_k\)</span> denotes the empirical centered moments.
This method can be performed by setting the argument <code>method</code> to <code>"mme"</code> in the call to <code>fitdist</code>.
The estimate is computed by a closed-form formula for the following distributions: normal, lognormal, exponential, Poisson, gamma, logistic, negative binomial, geometric, beta and uniform distributions.
In this case, for distributions characterized by one parameter (geometric, Poisson and exponential), this parameter is simply estimated by matching theoretical and observed means, and for distributions characterized by two parameters, these parameters are estimated by matching theoretical and observed means and variances <span class="citation">(<a href="#ref-Vose10" role="doc-biblioref">Vose 2010</a>)</span>.
For other distributions, the equation of moments is solved numerically using the <code>optim</code> function by minimizing the sum of squared differences between observed and theoretical moments (see the <strong>fitdistrplus</strong> reference manual for technical details <span class="citation">(<a href="#ref-fitdistrplus" role="doc-biblioref">Delignette-Muller et al. 2014</a>)</span>).</p>
<p>A classical data set from the Danish insurance industry published in <span class="citation">McNeil (<a href="#ref-mcneil97" role="doc-biblioref">1997</a>)</span> will be used to illustrate this method.
In <strong>fitdistrplus</strong>, the data set is stored in <code>danishuni</code> for the univariate version and contains the loss amounts collected at Copenhagen Reinsurance between 1980 and 1990.
In actuarial science, it is standard to consider positive heavy-tailed distributions and have a special focus on the right-tail of the distributions.
In this numerical experiment, we choose classic actuarial distributions for loss modelling: the lognormal distribution and the Pareto type II distribution <span class="citation">(<a href="#ref-Klugmanetal09" role="doc-biblioref">Klugman, Panjer, and Willmot 2009</a>)</span>.</p>
<p>The lognormal distribution is fitted to <code>danishuni</code> data set by matching moments implemented as a closed-form formula.
On the left-hand graph of Figure <a href="#fig:danishmme">3.2</a>, the fitted distribution functions obtained using the moment matching estimation (MME) and maximum likelihood estimation (MLE) methods are compared.
The MME method provides a more cautious estimation of the insurance risk as the MME-fitted distribution function (resp. MLE-fitted) underestimates (overestimates) the empirical distribution function for large values of claim amounts.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"danishuni"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    2167 obs. of  2 variables:</span></span>
<span><span class="co">##  $ Date: Date, format: "1980-01-03" "1980-01-04" ...</span></span>
<span><span class="co">##  $ Loss: num  1.68 2.09 1.73 1.78 4.61 ...</span></span></code></pre>
<p><br></p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fdanish.ln.MLE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"lnorm"</span><span class="op">)</span></span>
<span><span class="va">fdanish.ln.MME</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"lnorm"</span>, method <span class="op">=</span> <span class="st">"mme"</span>, order <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span><span class="op">)</span></span>
<span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://gitlab.com/vigou3/actuar" class="external-link">actuar</a></span><span class="op">)</span></span>
<span><span class="va">fdanish.P.MLE</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"pareto"</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">10</span>, scale <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, </span>
<span>                         lower <span class="op">=</span> <span class="fl">2</span><span class="op">+</span><span class="fl">1e-6</span>, upper <span class="op">=</span> <span class="cn">Inf</span><span class="op">)</span></span>
<span><span class="va">memp</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">x</span>, <span class="va">order</span><span class="op">)</span> <span class="fu"><a href="https://rdrr.io/r/base/sum.html" class="external-link">sum</a></span><span class="op">(</span><span class="va">x</span><span class="op">^</span><span class="va">order</span><span class="op">)</span> <span class="op">/</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">x</span><span class="op">)</span></span>
<span><span class="va">fdanish.P.MME</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"pareto"</span>, method <span class="op">=</span> <span class="st">"mme"</span>, order <span class="op">=</span> <span class="fl">1</span><span class="op">:</span><span class="fl">2</span>, memp <span class="op">=</span> <span class="st">"memp"</span>, </span>
<span>                         start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">10</span>, scale <span class="op">=</span> <span class="fl">10</span><span class="op">)</span>, lower <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span><span class="op">+</span><span class="fl">1e-6</span>, <span class="fl">2</span><span class="op">+</span><span class="fl">1e-6</span><span class="op">)</span>, </span>
<span>                         upper <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="cn">Inf</span>, <span class="cn">Inf</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fdanish.ln.MLE</span>, <span class="va">fdanish.ln.MME</span><span class="op">)</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lognormal MLE"</span>, <span class="st">"lognormal MME"</span><span class="op">)</span>,</span>
<span>        main <span class="op">=</span> <span class="st">"Fitting a lognormal distribution"</span>, xlogscale <span class="op">=</span> <span class="cn">TRUE</span>, datapch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fdanish.P.MLE</span>, <span class="va">fdanish.P.MME</span><span class="op">)</span>, legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Pareto MLE"</span>, <span class="st">"Pareto MME"</span><span class="op">)</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Fitting a Pareto distribution"</span>, xlogscale <span class="op">=</span> <span class="cn">TRUE</span>, datapch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:danishmme"></span>
<img src="fitdistrplus_vignette_files/figure-html/danishmme-1.png" alt="Comparison between MME and MLE when fitting a lognormal or a Pareto distribution to loss data from the `danishuni` data set."><p class="caption">
Figure 3.2: Comparison between MME and MLE when fitting a lognormal or a Pareto distribution to loss data from the <code>danishuni</code> data set.
</p>
</div>
<p><br></p>
<p>In a second time, a Pareto distribution, which gives more weight to the right-tail of the distribution, is fitted.
As the lognormal distribution, the Pareto has two parameters, which allows a fair comparison.</p>
<p>We use the implementation of the <strong>actuar</strong> package providing raw and centered moments for that distribution (in addition to <code>d</code>, <code>p</code>, <code>q</code> and <code>r</code> functions <span class="citation">(<a href="#ref-actuar12" role="doc-biblioref">Goulet 2012</a>)</span>.
Fitting a heavy-tailed distribution for which the first and the second moments do not exist for certain values of the shape parameter requires some cautiousness.
This is carried out by providing, for the optimization process, a lower and an upper bound for each parameter.
The code below calls the L-BFGS-B optimization method in <code>optim</code>, since this quasi-Newton allows box constraints <a class="footnote-ref" tabindex="0" data-bs-toggle="popover" data-bs-content="&lt;p&gt;That is what the B stands for.&lt;/p&gt;"><sup>2</sup></a>.
We choose match moments defined in Equation <a href="#eq:eq4">(3.1)</a>, and so a function for computing the empirical raw moment (called <code>memp</code> in our example) is passed to <code>fitdist</code>.
For two-parameter distributions (i.e., <span class="math inline">\(d=2\)</span>), Equations <a href="#eq:eq4">(3.1)</a> and <a href="#eq:eq5">(3.2)</a> are equivalent.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gofstat.html">gofstat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fdanish.ln.MLE</span>, <span class="va">fdanish.P.MLE</span>, <span class="va">fdanish.ln.MME</span>, <span class="va">fdanish.P.MME</span><span class="op">)</span>, </span>
<span>        fitnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lnorm.mle"</span>, <span class="st">"Pareto.mle"</span>, <span class="st">"lnorm.mme"</span>, <span class="st">"Pareto.mme"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Goodness-of-fit statistics</span></span>
<span><span class="co">##                              lnorm.mle Pareto.mle lnorm.mme Pareto.mme</span></span>
<span><span class="co">## Kolmogorov-Smirnov statistic    0.1375     0.3124    0.4368       0.37</span></span>
<span><span class="co">## Cramer-von Mises statistic     14.7911    37.7227   88.9503      55.43</span></span>
<span><span class="co">## Anderson-Darling statistic     87.1933   208.3388  416.2567     281.58</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Goodness-of-fit criteria</span></span>
<span><span class="co">##                                lnorm.mle Pareto.mle lnorm.mme Pareto.mme</span></span>
<span><span class="co">## Akaike's Information Criterion      8120       9250      9792       9409</span></span>
<span><span class="co">## Bayesian Information Criterion      8131       9261      9803       9420</span></span></code></pre>
<p>As shown on Figure <a href="#fig:danishmme">3.2</a>, MME and MLE fits are far less distant (when looking at the right-tail) for the Pareto distribution than for the lognormal distribution on this data set.
Furthermore, for these two distributions, the MME method better fits the right-tail of the distribution from a visual point of view.
This seems logical since empirical moments are influenced by large observed values.
In the previous traces, we gave the values of goodness-of-fit statistics.
Whatever the statistic considered, the MLE-fitted lognormal always provides the best fit to the observed data.</p>
<p>Maximum likelihood and moment matching estimations are certainly the most commonly used method for fitting distributions <span class="citation">(<a href="#ref-Cullen99" role="doc-biblioref">Cullen and Frey 1999</a>)</span>.
Keeping in mind that these two methods may produce very different results, the user should be aware of its great sensitivity to outliers when choosing the moment matching estimation.
This may be seen as an advantage in our example if the objective is to better describe the right tail of the distribution, but it may be seen as a drawback if the objective is different.</p>
<p>Fitting of a parametric distribution may also be done by matching theoretical quantiles of the parametric distributions (for specified probabilities) against the empirical quantiles <span class="citation">(<a href="#ref-Tse2009" role="doc-biblioref">Tse 2009</a>)</span>.
The equality of theoretical and empirical quantiles is expressed by Equation <a href="#eq:eq6">(3.3)</a> below, which is very similar to Equations <a href="#eq:eq4">(3.1)</a> and <a href="#eq:eq5">(3.2)</a>:</p>
<p><span class="math display" id="eq:eq6">\[\begin{equation}
F^{-1}(p_{k}|\theta)=Q_{n,p_{k}}\tag{3.3}
\end{equation}\]</span>
for <span class="math inline">\(k=1,\ldots,d\)</span>, with <span class="math inline">\(d\)</span> the number of parameters to estimate (dimension of <span class="math inline">\(\theta\)</span> if there is no fixed parameters) and <span class="math inline">\(Q_{n,p_{k}}\)</span> the empirical quantiles calculated from data for specified probabilities <span class="math inline">\(p_{k}\)</span>.</p>
<p>Quantile matching estimation (QME) is performed by setting the argument <code>method</code> to <code>"qme"</code> in the call to <code>fitdist</code> and adding an argument <code>probs</code> defining the probabilities for which the quantile matching is performed (see Figure <a href="#fig:danishqme">3.3</a>).
The length of this vector must be equal to the number of parameters to estimate (as the vector of moment orders for MME).
Empirical quantiles are computed using the <code>quantile</code> function of the <strong>stats</strong> package using <code>type=7</code> by default (see <code><a href="../reference/quantile.html">?quantile</a></code> and <span class="citation">Hyndman and Fan (<a href="#ref-hyndmanfan96" role="doc-biblioref">1996</a>)</span>).
But the type of quantile can be easily changed by using the <code>qty</code> argument in the call to the <code>qme</code> function.<br>
The quantile matching is carried out numerically, by minimizing the sum of squared differences between observed and theoretical quantiles.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fdanish.ln.QME1</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"lnorm"</span>, method <span class="op">=</span> <span class="st">"qme"</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span><span class="op">/</span><span class="fl">3</span>, <span class="fl">2</span><span class="op">/</span><span class="fl">3</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">fdanish.ln.QME2</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">danishuni</span><span class="op">$</span><span class="va">Loss</span>, <span class="st">"lnorm"</span>, method <span class="op">=</span> <span class="st">"qme"</span>, probs <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">8</span><span class="op">/</span><span class="fl">10</span>, <span class="fl">9</span><span class="op">/</span><span class="fl">10</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fdanish.ln.MLE</span>, <span class="va">fdanish.ln.QME1</span>, <span class="va">fdanish.ln.QME2</span><span class="op">)</span>, </span>
<span>        legend <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"MLE"</span>, <span class="st">"QME(1/3, 2/3)"</span>, <span class="st">"QME(8/10, 9/10)"</span><span class="op">)</span>, </span>
<span>        main <span class="op">=</span> <span class="st">"Fitting a lognormal distribution"</span>, xlogscale <span class="op">=</span> <span class="cn">TRUE</span>, datapch <span class="op">=</span> <span class="fl">20</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:danishqme"></span>
<img src="fitdistrplus_vignette_files/figure-html/danishqme-1.png" alt="Comparison between QME and MLE when fitting a lognormal distribution to loss data from the `danishuni` data set."><p class="caption">
Figure 3.3: Comparison between QME and MLE when fitting a lognormal distribution to loss data from the <code>danishuni</code> data set.
</p>
</div>
<p><br></p>
<p>Above is an example of fitting of a lognormal distribution to `danishuni} data set by matching probabilities <span class="math inline">\((p_1= 1/3, p_2=2/3)\)</span> and <span class="math inline">\((p_1= 8/10, p_2=9/10)\)</span>.
As expected, the second QME fit gives more weight to the right-tail of the distribution.
Compared to the maximum likelihood estimation, the second QME fit best suits the right-tail of the distribution, whereas the first QME fit best models the body of the distribution.
The quantile matching estimation is of particular interest when we need to focus around particular quantiles, e.g., <span class="math inline">\(p=99.5\%\)</span> in the Solvency II insurance context or <span class="math inline">\(p=5\%\)</span> for the HC5 estimation in the ecotoxicology context.</p>
</div>
<div class="section level3" number="3.2">
<h3 id="Customization">
<span class="header-section-number">3.2</span> Customization of the optimization algorithm<a class="anchor" aria-label="anchor" href="#Customization"></a>
</h3>
<p>Each time a numerical minimization is carried out in the <code>fitdistrplus</code> package, the <code>optim</code> function of the <strong>stats</strong> package is used by default with the <code>Nelder-Mead</code> method for distributions characterized by more than one parameter and the <code>BFGS</code> method for distributions characterized by only one parameter.
Sometimes the default algorithm fails to converge.
It is then interesting to change some options of the <code>optim</code> function or to use another optimization function than <code>optim</code> to minimize the objective function.
The argument <code>optim.method</code> can be used in the call to <code>fitdist</code> or <code>fitdistcens</code>.
It will internally be passed to <code>mledist</code>, <code>mmedist</code>, <code>mgedist</code> or <code>qmedist</code>, and to <code>optim</code> (see <code><a href="https://rdrr.io/r/stats/optim.html" class="external-link">?optim</a></code> for details about the different algorithms available).</p>
<p>Even if no error is raised when computing the optimization, changing the algorithm is of particular interest to enforce bounds on some parameters.
For instance, a volatility parameter <span class="math inline">\(\sigma\)</span> is strictly positive <span class="math inline">\(\sigma&gt;0\)</span> and a probability parameter <span class="math inline">\(p\)</span> lies in <span class="math inline">\(p\in [0,1]\)</span>.
This is possible by using arguments <code>lower</code> and/or <code>upper</code>, for which their use automatically forces <code>optim.method="L-BFGS-B"</code>.</p>
<p>Below are examples of fits of a gamma distribution <span class="math inline">\(\mathcal{G}(\alpha, \lambda)\)</span> to the <code>groundbeef</code> data set with various algorithms.
Note that the conjugate gradient algorithm (<code>CG</code>) needs far more iterations to converge (around 2500 iterations) compared to other algorithms (converging in less than 100 iterations).</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"groundbeef"</span><span class="op">)</span></span>
<span><span class="va">fNM</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span>, optim.method <span class="op">=</span> <span class="st">"Nelder-Mead"</span><span class="op">)</span></span>
<span><span class="va">fBFGS</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span>, optim.method <span class="op">=</span> <span class="st">"BFGS"</span><span class="op">)</span> </span>
<span><span class="va">fSANN</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span>, optim.method <span class="op">=</span> <span class="st">"SANN"</span><span class="op">)</span></span>
<span><span class="va">fCG</span> <span class="op">&lt;-</span> <span class="kw"><a href="https://rdrr.io/r/base/try.html" class="external-link">try</a></span><span class="op">(</span><span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span>, optim.method <span class="op">=</span> <span class="st">"CG"</span>,</span>
<span>                   control <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>maxit <span class="op">=</span> <span class="fl">10000</span><span class="op">)</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="kw">if</span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/class.html" class="external-link">inherits</a></span><span class="op">(</span><span class="va">fCG</span>, <span class="st">"try-error"</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span><span class="va">fCG</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>estimate <span class="op">=</span> <span class="cn">NA</span><span class="op">)</span><span class="op">}</span></span></code></pre></div>
<p>It is also possible to use another function than <code>optim</code> to minimize the objective function by specifying by the argument <code>custom.optim</code> in the call to <code>fitdist</code>.
It may be necessary to customize this optimization function to meet the following requirements.
(1) <code>custom.optim</code> function must have the following arguments: <code>fn</code> for the function to be optimized and <code>par</code> for the initialized parameters.
(2) <code>custom.optim</code> should carry out a MINIMIZATION and must return the following components: <code>par</code> for the estimate, <code>convergence</code> for the convergence code, <code>value=fn(par)</code> and <code>hessian</code>.
Below is an example of code written to wrap the <code>genoud</code> function from the <strong>rgenoud</strong> package in order to respect our optimization ``template’’.
The <strong>rgenoud</strong> package implements the genetic (stochastic) algorithm.</p>
<div class="sourceCode" id="cb38"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">mygenoud</span> <span class="op">&lt;-</span> <span class="kw">function</span><span class="op">(</span><span class="va">fn</span>, <span class="va">par</span>, <span class="va">...</span><span class="op">)</span> </span>
<span><span class="op">{</span></span>
<span>   <span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="https://github.com/JasjeetSekhon/rgenoud" class="external-link">rgenoud</a></span><span class="op">)</span></span>
<span>   <span class="va">res</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/pkg/rgenoud/man/genoud.html" class="external-link">genoud</a></span><span class="op">(</span><span class="va">fn</span>, starting.values <span class="op">=</span> <span class="va">par</span>, <span class="va">...</span><span class="op">)</span>        </span>
<span>   <span class="va">standardres</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">res</span>, convergence <span class="op">=</span> <span class="fl">0</span><span class="op">)</span></span>
<span>   <span class="kw"><a href="https://rdrr.io/r/base/function.html" class="external-link">return</a></span><span class="op">(</span><span class="va">standardres</span><span class="op">)</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The customized optimization function can then be passed as the argument <code>custom.optim</code> in the call to <code>fitdist</code> or <code>fitdistcens</code>.
The following code can for example be used to fit a gamma distribution to the <code>groundbeef</code> data set.
Note that in this example various arguments are also passed from <code>fitdist</code> to <code>genoud</code>: <code>nvars</code>, <code>Domains</code>, <code>boundary.enforcement</code>, <code>print.level</code> and <code>hessian</code>.
The code below compares all the parameter estimates (<span class="math inline">\(\hat\alpha\)</span>, <span class="math inline">\(\hat\lambda\)</span>) by the different algorithms: shape <span class="math inline">\(\alpha\)</span> and rate <span class="math inline">\(\lambda\)</span> parameters are relatively similar on this example, roughly 4.00 and 0.05, respectively.</p>
<div class="sourceCode" id="cb39"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fgenoud</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/mledist.html">mledist</a></span><span class="op">(</span><span class="va">groundbeef</span><span class="op">$</span><span class="va">serving</span>, <span class="st">"gamma"</span>, custom.optim <span class="op">=</span> <span class="va">mygenoud</span>, nvars <span class="op">=</span> <span class="fl">2</span>, </span>
<span>                   max.generations <span class="op">=</span> <span class="fl">10</span>, Domains <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0</span>, <span class="fl">0</span><span class="op">)</span>, <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">10</span>, <span class="fl">10</span><span class="op">)</span><span class="op">)</span>, </span>
<span>                   boundary.enforcement <span class="op">=</span> <span class="fl">1</span>, hessian <span class="op">=</span> <span class="cn">TRUE</span>, print.level <span class="op">=</span> <span class="fl">0</span>, P9 <span class="op">=</span> <span class="fl">10</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Loading required package: rgenoud</span></span></code></pre>
<pre><code><span><span class="co">## ##  rgenoud (Version 5.9-0.3, Build Date: 2022-04-19)</span></span>
<span><span class="co">## ##  See http://sekhon.berkeley.edu/rgenoud for additional documentation.</span></span>
<span><span class="co">## ##  Please cite software as:</span></span>
<span><span class="co">## ##   Walter Mebane, Jr. and Jasjeet S. Sekhon. 2011.</span></span>
<span><span class="co">## ##   ``Genetic Optimization Using Derivatives: The rgenoud package for R.''</span></span>
<span><span class="co">## ##   Journal of Statistical Software, 42(11): 1-26. </span></span>
<span><span class="co">## ##</span></span></code></pre>
<div class="sourceCode" id="cb42"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/cbind.html" class="external-link">cbind</a></span><span class="op">(</span>NM <span class="op">=</span> <span class="va">fNM</span><span class="op">$</span><span class="va">estimate</span>, BFGS <span class="op">=</span> <span class="va">fBFGS</span><span class="op">$</span><span class="va">estimate</span>, SANN <span class="op">=</span> <span class="va">fSANN</span><span class="op">$</span><span class="va">estimate</span>, CG <span class="op">=</span> <span class="va">fCG</span><span class="op">$</span><span class="va">estimate</span>, </span>
<span>      fgenoud <span class="op">=</span> <span class="va">fgenoud</span><span class="op">$</span><span class="va">estimate</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">##            NM    BFGS    SANN      CG fgenoud</span></span>
<span><span class="co">## shape 4.00825 4.22848 3.96743 4.12850 4.00834</span></span>
<span><span class="co">## rate  0.05442 0.05742 0.05385 0.05606 0.05443</span></span></code></pre>
</div>
<div class="section level3" number="3.3">
<h3 id="otherdata">
<span class="header-section-number">3.3</span> Fitting distributions to other types of data<a class="anchor" aria-label="anchor" href="#otherdata"></a>
</h3>
<p><em>This section was modified since the publication of this vignette in the Journal of Statistical Software in order to include new goodness-of-fit plots for censored and discrete data.</em></p>
<p>Analytical methods often lead to semi-quantitative results which are referred to as censored data.
Observations only known to be under a limit of detection are left-censored data.
Observations only known to be above a limit of quantification are right-censored data.
Results known to lie between two bounds are interval-censored data.
These two bounds may correspond to a limit of detection and a limit of quantification, or more generally to uncertainty bounds around the observation.
Right-censored data are also commonly encountered with survival data <span class="citation">(<a href="#ref-kleinmoeschberger03" role="doc-biblioref">Klein and Moeschberger 2003</a>)</span>.
A data set may thus contain right-, left-, or interval-censored data, or may be a mixture of these categories, possibly with different upper and lower bounds.
Censored data are sometimes excluded from the data analysis or replaced by a fixed value, which in both cases may lead to biased results.
A more recommended approach to correctly model such data is based upon maximum likelihood <span class="citation">Helsel (<a href="#ref-helsel05" role="doc-biblioref">2005</a>)</span>.</p>
<p>Censored data may thus contain left-censored, right-censored and interval-censored values, with several lower and upper bounds.
Before their use in package <strong>fitdistrplus</strong>, such data must be coded into a dataframe with two columns, respectively named <code>left</code> and <code>right</code>, describing each observed value as an interval.
The <code>left</code> column contains either <code>NA</code> for left censored observations, the left bound of the interval for interval censored observations, or the observed value for non-censored observations.
The <code>right</code> column contains either <code>NA</code> for right censored observations, the right bound of the interval for interval censored observations, or the observed value for non-censored observations.
To illustrate the use of package <strong>fitdistrplus</strong> to fit distributions to censored continous data, we will use another data set from ecotoxicology, included in our package and named <code>salinity</code>.
This data set contains acute salinity tolerance (LC50 values in electrical conductivity, <span class="math inline">\(mS\)</span>.<span class="math inline">\(cm^{-1}\)</span>) of riverine macro-invertebrates taxa from the southern Murray-Darling Basin in Central Victoria, Australia <span class="citation">(<a href="#ref-kefford07" role="doc-biblioref">Kefford et al. 2007</a>)</span>.</p>
<div class="sourceCode" id="cb44"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"salinity"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">salinity</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    108 obs. of  2 variables:</span></span>
<span><span class="co">##  $ left : num  20 20 20 20 20 21.5 15 20 23.7 25 ...</span></span>
<span><span class="co">##  $ right: num  NA NA NA NA NA 21.5 30 25 23.7 NA ...</span></span></code></pre>
<p>Using censored data such as those coded in the <code>salinity} data set, the empirical distribution can be plotted using the</code>plotdistcens} function.
In older versions of the package, by default this function used the Expectation-Maximization approach of <span class="citation">Turnbull (<a href="#ref-Turnbull74" role="doc-biblioref">1974</a>)</span> to compute the overall empirical cdf curve with optional confidence intervals, by calls to <code>survfit</code> and <code>plot.survfit</code> functions from the <strong>survival</strong> package.
Even if this representation is always available (by fixing the argument <code>NPMLE.method</code> to <code>"Turnbull.middlepoints"</code>), now the default plot of the empirical cumulative distribution function (ECDF) explicitly represents the regions of non uniqueness of the NPMLE ECDF.
The default computation of those regions of non uniqueness and their associated masses uses the non parametric maximum likelihood estimation (NPMLE) approach developped by Wang <span class="citation">Wang and Fani (<a href="#ref-Wang2018" role="doc-biblioref">2018</a>)</span>.<br>
Figure <a href="#fig:cdfcompcens">3.5</a> shows on the top left the new plot of data together with two fitted distributions.
Grey filled rectangles in such a plot represent the regions of non uniqueness of the NPMLE ECDF.</p>
<p>A less rigorous but sometimes more illustrative plot can be obtained by fixing the argument <code>NPMLE</code> to <code>FALSE</code> in the call to <code>plotdistcens</code>
(see Figure <a href="#fig:plotsalinity2">3.4</a> for an example and the help page of Function <code>plotdistcens</code> for details).
This plot enables to see the real nature of censored data, as points and intervals, but the difficulty in building such a plot is to define a relevant ordering of observations.</p>
<div class="sourceCode" id="cb46"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/plotdistcens.html">plotdistcens</a></span><span class="op">(</span><span class="va">salinity</span>, NPMLE <span class="op">=</span> <span class="cn">FALSE</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:plotsalinity2"></span>
<img src="fitdistrplus_vignette_files/figure-html/plotsalinity2-1.png" alt="Simple plot of censored raw data (72-hour acute salinity tolerance of riverine macro-invertebrates from the `salinity` data set) as ordered points and intervals."><p class="caption">
Figure 3.4: Simple plot of censored raw data (72-hour acute salinity tolerance of riverine macro-invertebrates from the <code>salinity</code> data set) as ordered points and intervals.
</p>
</div>
<p><br></p>
<p>As for non censored data, one or more parametric distributions can be fitted to the censored data set, one at a time, but using in this case the <code>fitdistcens</code> function.
This function estimates the vector of distribution parameters <span class="math inline">\(\theta\)</span> by maximizing the likelihood for censored data defined as:</p>
<p><span class="math display" id="eq:eq7">\[\begin{equation}
L(\theta) = \prod_{i=1}^{N_{nonC}} f(x_{i}|\theta)\times \prod_{j=1}^{N_{leftC}} F(x^{upper}_{j}|\theta) \\
\times \prod_{k=1}^{N_{rightC}} (1- F(x^{lower}_{k}|\theta))\times \prod_{m=1}^{N_{intC}} (F(x^{upper}_{m}|\theta)- F(x^{lower}_{j}|\theta))\tag{3.4}
\end{equation}\]</span></p>
<p>with <span class="math inline">\(x_{i}\)</span> the <span class="math inline">\(N_{nonC}\)</span> non-censored observations,
<span class="math inline">\(x^{upper}_{j}\)</span> upper values defining the <span class="math inline">\(N_{leftC}\)</span> left-censored observations,
<span class="math inline">\(x^{lower}_{k}\)</span> lower values defining the <span class="math inline">\(N_{rightC}\)</span> right-censored observations,
<span class="math inline">\([x^{lower}_{m} ; x^{upper}_{m}]\)</span> the intervals defining the <span class="math inline">\(N_{intC}\)</span> interval-censored observations, and F the cumulative distribution function of the parametric distribution <span class="citation">Helsel (<a href="#ref-helsel05" role="doc-biblioref">2005</a>)</span>.</p>
<p>As <code>fitdist</code>, <code>fitdistcens</code> returns the results of the fit of any parametric distribution to a data set as an S3 class object that can be easily printed, summarized or plotted.
For the <code>salinity</code> data set, a lognormal distribution or a loglogistic can be fitted as commonly done in ecotoxicology for such data.
As with <code>fitdist</code>, for some distributions (see <span class="citation">Delignette-Muller et al. (<a href="#ref-fitdistrplus" role="doc-biblioref">2014</a>)</span> for details), it is necessary to specify initial values for the distribution parameters in the argument <code>start</code>.
The <code>plotdistcens</code> function can help to find correct initial values for the distribution parameters in non trivial cases, by a manual iterative use if necessary.</p>
<div class="sourceCode" id="cb47"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fsal.ln</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdistcens.html">fitdistcens</a></span><span class="op">(</span><span class="va">salinity</span>, <span class="st">"lnorm"</span><span class="op">)</span></span>
<span><span class="va">fsal.ll</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdistcens.html">fitdistcens</a></span><span class="op">(</span><span class="va">salinity</span>, <span class="st">"llogis"</span>, start <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span>shape <span class="op">=</span> <span class="fl">5</span>, scale <span class="op">=</span> <span class="fl">40</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fsal.ln</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fitting of the distribution ' lnorm ' By maximum likelihood on censored data </span></span>
<span><span class="co">## Parameters</span></span>
<span><span class="co">##         estimate Std. Error</span></span>
<span><span class="co">## meanlog   3.3854    0.06487</span></span>
<span><span class="co">## sdlog     0.4961    0.05455</span></span>
<span><span class="co">## Loglikelihood:  -139.1   AIC:  282.1   BIC:  287.5 </span></span>
<span><span class="co">## Correlation matrix:</span></span>
<span><span class="co">##         meanlog  sdlog</span></span>
<span><span class="co">## meanlog  1.0000 0.2938</span></span>
<span><span class="co">## sdlog    0.2938 1.0000</span></span></code></pre>
<div class="sourceCode" id="cb49"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/summary.html" class="external-link">summary</a></span><span class="op">(</span><span class="va">fsal.ll</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fitting of the distribution ' llogis ' By maximum likelihood on censored data </span></span>
<span><span class="co">## Parameters</span></span>
<span><span class="co">##       estimate Std. Error</span></span>
<span><span class="co">## shape    3.421     0.4158</span></span>
<span><span class="co">## scale   29.930     1.9447</span></span>
<span><span class="co">## Loglikelihood:  -140.1   AIC:  284.1   BIC:  289.5 </span></span>
<span><span class="co">## Correlation matrix:</span></span>
<span><span class="co">##         shape   scale</span></span>
<span><span class="co">## shape  1.0000 -0.2022</span></span>
<span><span class="co">## scale -0.2022  1.0000</span></span></code></pre>
<p>Computations of goodness-of-fit statistics have not yet been developed for fits using censored data but the quality of fit can be judged using Akaike and Schwarz’s Bayesian information criteria (AIC and BIC) and the goodness-of-fit CDF plot, respectively provided when summarizing or plotting an object of class <code>fitdistcens</code>.
Functions <code>cdfcompcens</code>, <code>qqcompcens</code> and <code>ppcompcens</code> can also be used to compare the fit of various distributions to the same censored data set.
Their calls are similar to the ones of <code>cdfcomp</code>, <code>qqcomp</code> and <code>ppcomp</code>.
Below are examples of use of those functions with the two fitted distributions to the <code>salinity</code> data set (see Figure <a href="#fig:cdfcompcens">3.5</a>).
When <code>qqcompcens</code> and <code>ppcompcens</code> are used with more than one fitted distribution, the non uniqueness rectangles are not filled and a small noise is added on the y-axis in order to help the visualization of various fits.
But we rather recommend the use of the <code>plotstyle</code> <code>ggplot</code> of <code>qqcompcens</code> and <code>ppcompcens</code> to compare the fits of various distributions as it provides a clearer plot splitted in facets (see <code><a href="../reference/graphcompcens.html">?graphcompcens</a></code>).</p>
<div class="sourceCode" id="cb51"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">2</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcompcens.html">cdfcompcens</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fsal.ln</span>, <span class="va">fsal.ll</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lognormal"</span>, <span class="st">"loglogistic "</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcompcens.html">qqcompcens</a></span><span class="op">(</span><span class="va">fsal.ln</span>, legendtext <span class="op">=</span> <span class="st">"lognormal"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcompcens.html">ppcompcens</a></span><span class="op">(</span><span class="va">fsal.ln</span>, legendtext <span class="op">=</span> <span class="st">"lognormal"</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcompcens.html">qqcompcens</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">fsal.ln</span>, <span class="va">fsal.ll</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"lognormal"</span>, <span class="st">"loglogistic "</span><span class="op">)</span>,</span>
<span>           main <span class="op">=</span> <span class="st">"Q-Q plot with 2 dist."</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:cdfcompcens"></span>
<img src="fitdistrplus_vignette_files/figure-html/cdfcompcens-1.png" alt="Some goodness-of-fit plots for fits of a lognormal and a loglogistic distribution to censored data: LC50 values from the `salinity` data set."><p class="caption">
Figure 3.5: Some goodness-of-fit plots for fits of a lognormal and a loglogistic distribution to censored data: LC50 values from the <code>salinity</code> data set.
</p>
</div>
<p><br></p>
<p>Function <code>bootdistcens</code> is the equivalent of <code>bootdist</code> for censored data, except that it only proposes nonparametric bootstrap.
Indeed, it is not obvious to simulate censoring within a parametric bootstrap resampling procedure.
The generic function <code>quantile</code> can also be applied to an object of class <code>fitdistcens</code> or <code>bootdistcens</code>, as for continuous non-censored data.</p>
<p>In addition to the fit of distributions to censored or non censored continuous data, our package can also accomodate discrete variables, such as count numbers, using the functions developped for continuous non-censored data.
These functions will provide somewhat different graphs and statistics, taking into account the discrete nature of the modeled variable.
The discrete nature of the variable is automatically recognized when a classical distribution is fitted to data (binomial, negative binomial, geometric, hypergeometric and Poisson distributions) but must be indicated by fixing argument <code>discrete</code> to <code>TRUE</code> in the call to functions in other cases.
The <code>toxocara</code> data set included in the package corresponds to the observation of such a discrete variable.
Numbers of <em>Toxocara cati</em> parasites present in digestive tract are reported from a random sampling of feral cats living on Kerguelen island <span class="citation">(<a href="#ref-Fromont01" role="doc-biblioref">Fromont et al. 2001</a>)</span>.
We will use it to illustrate the case of discrete data.</p>
<div class="sourceCode" id="cb52"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/utils/data.html" class="external-link">data</a></span><span class="op">(</span><span class="st">"toxocara"</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/str.html" class="external-link">str</a></span><span class="op">(</span><span class="va">toxocara</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## 'data.frame':    53 obs. of  1 variable:</span></span>
<span><span class="co">##  $ number: int  0 0 0 0 0 0 0 0 0 0 ...</span></span></code></pre>
<p>The fit of a discrete distribution to discrete data by maximum likelihood estimation requires the same procedure as for continuous non-censored data.
As an example, using the <code>toxocara</code> data set, Poisson and negative binomial distributions can be easily fitted.</p>
<div class="sourceCode" id="cb54"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ftoxo.P</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">toxocara</span><span class="op">$</span><span class="va">number</span>, <span class="st">"pois"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fitting of the distribution ' pois ' by maximum likelihood </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##        estimate Std. Error</span></span>
<span><span class="co">## lambda    8.679     0.4047</span></span></code></pre>
<div class="sourceCode" id="cb56"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="op">(</span><span class="va">ftoxo.nb</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/fitdist.html">fitdist</a></span><span class="op">(</span><span class="va">toxocara</span><span class="op">$</span><span class="va">number</span>, <span class="st">"nbinom"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Fitting of the distribution ' nbinom ' by maximum likelihood </span></span>
<span><span class="co">## Parameters:</span></span>
<span><span class="co">##      estimate Std. Error</span></span>
<span><span class="co">## size   0.3971    0.08289</span></span>
<span><span class="co">## mu     8.6803    1.93501</span></span></code></pre>
<p>For discrete distributions, the plot of an object of class <code>fitdist</code> simply provides two goodness-of-fit plots comparing empirical and theoretical distributions in density and in CDF.
Functions <code>cdfcomp</code> and <code>denscomp</code> can also be used to compare several plots to the same data set, as follows for the previous fits (Figure <a href="#fig:fittoxocarapoisnbinom">3.6</a>).</p>
<div class="sourceCode" id="cb58"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/graphics/par.html" class="external-link">par</a></span><span class="op">(</span>mfrow <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">1</span>, <span class="fl">2</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">denscomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ftoxo.P</span>, <span class="va">ftoxo.nb</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Poisson"</span>, <span class="st">"negative binomial"</span><span class="op">)</span>, fitlty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/graphcomp.html">cdfcomp</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ftoxo.P</span>, <span class="va">ftoxo.nb</span><span class="op">)</span>, legendtext <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Poisson"</span>, <span class="st">"negative binomial"</span><span class="op">)</span>, fitlty <span class="op">=</span> <span class="fl">1</span><span class="op">)</span></span></code></pre></div>
<div class="figure" style="text-align: center">
<span style="display:block;" id="fig:fittoxocarapoisnbinom"></span>
<img src="fitdistrplus_vignette_files/figure-html/fittoxocarapoisnbinom-1.png" alt="Comparison of the fits of a negative binomial and a Poisson distribution to numbers of *Toxocara cati* parasites from the `toxocara` data set."><p class="caption">
Figure 3.6: Comparison of the fits of a negative binomial and a Poisson distribution to numbers of <em>Toxocara cati</em> parasites from the <code>toxocara</code> data set.
</p>
</div>
<p><br></p>
<p>When fitting discrete distributions, the Chi-squared statistic is computed by the <code>gofstat</code> function using cells defined by the argument <code>chisqbreaks</code> or cells automatically defined from the data in order to reach roughly the same number of observations per cell.
This number is roughly equal to the argument <code>meancount</code>, or sligthly greater if there are some ties.
The choice to define cells from the empirical distribution (data), and not from the theoretical distribution, was done to enable the comparison of Chi-squared values obtained with different distributions fitted on a same data set.
If arguments <code>chisqbreaks</code> and <code>meancount</code> are both omitted, <code>meancount</code> is fixed in order to obtain roughly <span class="math inline">\((4n)^{2/5}\)</span> cells, with <span class="math inline">\(n\)</span> the length of the data set <span class="citation">(<a href="#ref-Vose10" role="doc-biblioref">Vose 2010</a>)</span>.
Using this default option the two previous fits are compared as follows, giving the preference to the negative binomial distribution, from both Chi-squared statistics and information criteria:</p>
<div class="sourceCode" id="cb59"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/gofstat.html">gofstat</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/list.html" class="external-link">list</a></span><span class="op">(</span><span class="va">ftoxo.P</span>, <span class="va">ftoxo.nb</span><span class="op">)</span>, fitnames <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"Poisson"</span>, <span class="st">"negative binomial"</span><span class="op">)</span><span class="op">)</span></span></code></pre></div>
<pre><code><span><span class="co">## Chi-squared statistic:  31257 7.486 </span></span>
<span><span class="co">## Degree of freedom of the Chi-squared distribution:  5 4 </span></span>
<span><span class="co">## Chi-squared p-value:  0 0.1123 </span></span>
<span><span class="co">##    the p-value may be wrong with some theoretical counts &lt; 5  </span></span>
<span><span class="co">## Chi-squared table:</span></span>
<span><span class="co">##       obscounts theo Poisson theo negative binomial</span></span>
<span><span class="co">## &lt;= 0         14     0.009014                 15.295</span></span>
<span><span class="co">## &lt;= 1          8     0.078237                  5.809</span></span>
<span><span class="co">## &lt;= 3          6     1.321767                  6.845</span></span>
<span><span class="co">## &lt;= 4          6     2.131298                  2.408</span></span>
<span><span class="co">## &lt;= 9          6    29.827829                  7.835</span></span>
<span><span class="co">## &lt;= 21         6    19.626224                  8.271</span></span>
<span><span class="co">## &gt; 21          7     0.005631                  6.537</span></span>
<span><span class="co">## </span></span>
<span><span class="co">## Goodness-of-fit criteria</span></span>
<span><span class="co">##                                Poisson negative binomial</span></span>
<span><span class="co">## Akaike's Information Criterion    1017             322.7</span></span>
<span><span class="co">## Bayesian Information Criterion    1019             326.6</span></span></code></pre>
</div>
</div>
<div class="section level2" number="4">
<h2 id="ccl">
<span class="header-section-number">4</span> Conclusion<a class="anchor" aria-label="anchor" href="#ccl"></a>
</h2>
<p>The R package <strong>fitdistrplus</strong> allows to easily fit distributions.
Our main objective while developing this package was to provide tools for helping R users to fit distributions to data.
We have been encouraged to pursue our work by feedbacks from users of our package in various areas as food or environmental risk assessment, epidemiology, ecology, molecular biology, genomics, bioinformatics, hydraulics, mechanics, financial and actuarial mathematics or operations research.
Indeed, this package is already used by a lot of practionners and academics
for simple MLE fits <span class="citation">Voigt et al. (<a href="#ref-voigtetal14" role="doc-biblioref">2014</a>)</span>,
for MLE fits and goodness-of-fit statistics <span class="citation">Vaninsky (<a href="#ref-vaninsky13" role="doc-biblioref">2013</a>)</span>,
for MLE fits and bootstrap <span class="citation">Rigaux et al. (<a href="#ref-Rigaux2014" role="doc-biblioref">2014</a>)</span>,
for MLE fits, bootstrap and goodness-of-fit statistics <span class="citation">(<a href="#ref-larrasetal13" role="doc-biblioref">Larras, Montuelle, and Bouchez 2013</a>)</span>,
for MME fit <span class="citation">Sato et al. (<a href="#ref-satoetal13" role="doc-biblioref">2013</a>)</span>,
for censored MLE and bootstrap <span class="citation">Contreras, Huerta, and Arnold (<a href="#ref-contrerasetal2013" role="doc-biblioref">2013</a>)</span>,
for graphic analysing in <span class="citation">(<a href="#ref-anandetal12" role="doc-biblioref">Anand, Yeturu, and Chandra 2012</a>)</span>,
for grouped-data fitting methods <span class="citation">(<a href="#ref-fusteinercostafreda12" role="doc-biblioref">Fu, Steiner, and Costafreda 2012</a>)</span>
or more generally <span class="citation">Drake, Chalabi, and Coker (<a href="#ref-drakeetal2014" role="doc-biblioref">2014</a>)</span>.</p>
<p>The <strong>fitdistrplus</strong> package is complementary with the <strong>distrMod</strong> package <span class="citation">(<a href="#ref-distrModJSS" role="doc-biblioref">Kohl and Ruckdeschel 2010</a>)</span>.
<strong>distrMod</strong> provides an even more flexible way to estimate distribution parameters but its use requires a greater initial investment to learn how to manipulate the S4 classes and methods developed in the <code>distr</code>-family packages.</p>
<p>Many extensions of the <strong>fitdistrplus</strong> package are planned in the future: we target to extend to censored data some methods for the moment only available for non-censored data, especially concerning goodness-of-fit evaluation and fitting methods.
We will also enlarge the choice of fitting methods for non-censored data, by proposing new goodness-of-fit distances (e.g., distances based on quantiles) for maximum goodness-of-fit estimation and new types of moments (e.g., limited expected values) for moment matching estimation.
At last, we will consider the case of multivariate distribution fitting.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="acknowledgments">Acknowledgments<a class="anchor" aria-label="anchor" href="#acknowledgments"></a>
</h2>
<p>The package would not have been at this stage without the stimulating contribution of Régis Pouillot and Jean-Baptiste Denis, especially for its conceptualization.
We also want to thank Régis Pouillot for his very valuable comments on the first version of this paper.</p>
<p>The authors gratefully acknowledges the two anonymous referees and the Editor for useful and constructive comments.
The remaining errors, of course, should be attributed to the authors alone.</p>
<p><br></p>
</div>
<div class="section level2">
<h2 id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent">
<div id="ref-anandetal12" class="csl-entry">
Anand, P., K. Yeturu, and N. Chandra. 2012. <span>“PocketAnnotate: Towards Site-Based Function Annotation.”</span> <em>Nucleic Acids Research</em> 40: 1–9.
</div>
<div id="ref-bagariaetal12" class="csl-entry">
Bagaria, A., V. Jaravine, Y. J. Huang, G. T. Montelione, and P. Güntert. 2012. <span>“Protein Structure Validation by Generalized Linear Model Root-Mean-Square Deviation Prediction.”</span> <em>Protein Science</em> 21 (2): 229–38.
</div>
<div id="ref-benavidesetal12" class="csl-entry">
Benavides-Piccione, R., I. Fernaud-Espinosa, V. Robles, R. Yuste, and J. DeFelipe. 2012. <span>“Age-Based Comparison of Human Dendritic Spine Structure Using Complete Three-Dimensional Reconstructions.”</span> <em>Cerebral Cortex</em> 23 (8): 1798–1810.
</div>
<div id="ref-Blom" class="csl-entry">
Blom, G. 1959. <em>Statistical Estimates and Transformed Beta Variables</em>. 1st ed. John Wiley &amp; Sons.
</div>
<div id="ref-breitbach12" class="csl-entry">
Breitbach, N., K. Böhning-Gaese, I. Laube, and M. Schleuning. 2012. <span>“Short Seed-Dispersal Distances and Low Seedling Recruitment in Farmland Populations of Bird-Dispersed Cherry Trees.”</span> <em>Journal of Ecology</em> 100 (6): 1349–58.
</div>
<div id="ref-busschaertetal10" class="csl-entry">
Busschaert, P., A. H. Geeraerd, M. Uyttendaele, and J. F. VanImpe. 2010. <span>“Estimating Distributions Out of Qualitative and (Semi)quantitative Microbiological Contamination Data for Use in Risk Assessment.”</span> <em>International Journal of Food Microbiology</em> 138: 260–69.
</div>
<div id="ref-callauetal13" class="csl-entry">
Callau Poduje, Ana Claudia, Aslan Belli, and Uwe Haberlandt. 2013. <span>“Dam Risk Assessment Based on Univariate Versus Bivariate Statistical Approaches - a Case Study for Argentina.”</span> <em>Hydrological Sciences Journal</em>. <a href="https://doi.org/10.1080/02626667.2013.871014" class="external-link">https://doi.org/10.1080/02626667.2013.871014</a>.
</div>
<div id="ref-casellaberger02" class="csl-entry">
Casella, G., and R. L. Berger. 2002. <em>Statistical Inference</em>. 2nd ed. Duxbury Thomson Learning.
</div>
<div id="ref-commeauetal12" class="csl-entry">
Commeau, N., E. Parent, M.-L. Delignette-Muller, and M. Cornu. 2012. <span>“Fitting a Lognormal Distribution to Enumeration and Absence/Presence Data.”</span> <em>International Journal of Food Microbiology</em> 155: 146–52.
</div>
<div id="ref-contrerasetal2013" class="csl-entry">
Contreras, V. De La Huerta, H. Vaquera Huerta, and B. C. Arnold. 2013. <span>“A Test for Equality of Variance with Censored Samples.”</span> <em>Journal of Statistical Computation and Simulation</em>. <a href="https://doi.org/10.1080/00949655.2013.825095" class="external-link">https://doi.org/10.1080/00949655.2013.825095</a>.
</div>
<div id="ref-croucheretal12" class="csl-entry">
Croucher, N. J., S. R. Harris, L. Barquist, J. Parkhill, and S. D. Bentley. 2012. <span>“A High-Resolution View of Genome-Wide Pneumococcal Transformation.”</span> <em>PLoS Pathogens</em> 8 (6): e1002745.
</div>
<div id="ref-Cullen99" class="csl-entry">
Cullen, A. C., and H. C. Frey. 1999. <em>Probabilistic Techniques in Exposure Assessment</em>. 1st ed. Plenum Publishing Co.
</div>
<div id="ref-Stephens86" class="csl-entry">
D’Agostino, R. B., and M. A. Stephens. 1986. <em>Goodness-of-Fit Techniques</em>. 1st ed. Dekker.
</div>
<div id="ref-Daelmanetal13" class="csl-entry">
Daelman, Jeff, Jeanne-Marie Membré, Liesbeth Jacxsens, An Vermeulen, Frank Devlieghere, and Mieke Uyttendaele. 2013. <span>“A Quantitative Microbiological Exposure Assessment Model for <em>Bacillus Cereus</em> in REPFEDs.”</span> <em>International Journal of Food Microbiology</em> 166 (3): 433–49.
</div>
<div id="ref-Delignette08" class="csl-entry">
Delignette-Muller, M. L., and M. Cornu. 2008. <span>“Quantitative Risk Assessment for Escherichia Coli O157:H7 in Frozen Ground Beef Patties Consumed by Young Children in French Households.”</span> <em>International Journal of Food Microbiology</em> 128 (1): 158–64. https://doi.org/<a href="https://doi.org/10.1016/j.ijfoodmicro.2008.05.040" class="external-link">https://doi.org/10.1016/j.ijfoodmicro.2008.05.040</a>.
</div>
<div id="ref-fitdistrplus" class="csl-entry">
Delignette-Muller, M. L., R. Pouillot, J. B. Denis, and C. Dutang. 2014. <em>: Help to Fit of a Parametric Distribution to Non-Censored or Censored Data</em>. <a href="https://cran.r-project.org/package=fitdistrplus" class="external-link">https://cran.r-project.org/package=fitdistrplus</a>.
</div>
<div id="ref-drakeetal2014" class="csl-entry">
Drake, T., Z. Chalabi, and R. Coker. 2014. <span>“<span class="nocase">Buy Now, saved Later? The Critical Impact of Time-to-Pandemic Uncertainty on Pandemic Cost-Effectiveness Analyses</span>.”</span> <em>Health Policy and Planning</em>. <a href="https://doi.org/10.1093/heapol/czt101" class="external-link">https://doi.org/10.1093/heapol/czt101</a>.
</div>
<div id="ref-actuarJSS" class="csl-entry">
Dutang, C., V. Goulet, and M. Pigeon. 2008. <span>“<span class="nocase">: an R Package for Actuarial Science</span>.”</span> <em>Journal of Statistical Software</em> 25 (7): 1–37.
</div>
<div id="ref-efrontibshirani94" class="csl-entry">
Efron, B., and R. J. Tibshirani. 1994. <em>An Introduction to the Bootstrap</em>. 1st ed. Chapman &amp; Hall.
</div>
<div id="ref-eiketal13" class="csl-entry">
Eik, M., K. Luhmus, M. Tigasson, M. Listak, J. Puttonen, and H. Herrmann. 2013. <span>“<span class="nocase">DC-Conductivity Testing Combined with Photometry for Measuring Fibre Orientations in SFRC</span>.”</span> <em>Journal of Materials Science</em> 48 (10): 3745–59.
</div>
<div id="ref-eling12" class="csl-entry">
Eling, M. 2012. <span>“<span class="nocase">Fitting Insurance Claims to Skewed Distributions: Are the Skew-normal and the Skew-student Good Models?</span>”</span> <em>Insurance: Mathematics and Economics</em> 51 (2): 239–48.
</div>
<div id="ref-fiorellietal13" class="csl-entry">
Fiorelli, L. E., M. D. Ezcurra, E. M. Hechenleitner, E. Argañaraz, R. Jeremias, A. Taborda, M. J. Trotteyn, M. Belén von Baczko, and J. B. Desojo. 2013. <span>“<span class="nocase">The Oldest Known Communal Latrines Provide Evidence of Gregarism in Triassic Megaherbivores</span>.”</span> <em>Scientific Reports</em> 3 (3348): 1–7.
</div>
<div id="ref-Fromont01" class="csl-entry">
Fromont, E, L Morvilliers, M Artois, and D Pontier. 2001. <span>“<span class="nocase">Parasite Richness and Abundance in Insular and Mainland Feral Cats: Insularity or Density?</span>”</span> <em><span>Parasitology</span></em> <span>123</span> (<span>Part 2</span>): 143–51.
</div>
<div id="ref-fusteinercostafreda12" class="csl-entry">
Fu, C. H. Y., H. Steiner, and S. G. Costafreda. 2012. <span>“Predictive Neural Biomarkers of Clinical Response in Depression: A Meta-Analysis of Functional and Structural Neuroimaging Studies of Pharmacological and Psychological Therapies.”</span> <em>Neurobiology of Disease</em> 52: 75–83.
</div>
<div id="ref-varoetal12" class="csl-entry">
González-Varo, J. P., J. V. López-Bao, and J. Guitián. 2012. <span>“Functional Diversity Among Seed Dispersal Kernels Generated by Carnivorous Mammals.”</span> <em>Journal of Animal Ecology</em> <span>82</span>: 562–71.
</div>
<div id="ref-actuar12" class="csl-entry">
Goulet, V. 2012. <em>: An r Package for Actuarial Science</em>. <a href="https://cran.r-project.org/package=actuar" class="external-link">https://cran.r-project.org/package=actuar</a>.
</div>
<div id="ref-Guillier2013471" class="csl-entry">
Guillier, Laurent, Corinne Danan, Hélène Bergis, Marie-Laure Delignette-Muller, Sophie Granier, Sylvie Rudelle, Annie Beaufort, and Anne Brisabois. 2013. <span>“<span class="nocase">Use of Quantitative Microbial Risk Assessment when Investigating Foodborne Illness Outbreaks: the Example of a Monophasic <em>Salmonella Typhimurium</em> 4,5,12:i:- Outbreak Implicating Beef Burgers</span>.”</span> <em>International Journal of Food Microbiology</em> 166 (3): 471–78.
</div>
<div id="ref-helsel05" class="csl-entry">
Helsel, D. R. 2005. <em>Nondetects and Data Analysis: Statistics for Censored Environmental Data</em>. 1st ed. John Wiley &amp; Sons.
</div>
<div id="ref-hiranoetal94" class="csl-entry">
Hirano, S. S., M. K. Clayton, and C. D. Upper. 1994. <span>“<span class="nocase">Estimation of and Temporal Changes in Means and Variances of Populations of <em>Pseudomonas syringae</em> on Snap Bean Leaflets</span>.”</span> <em>Phytopathology</em> 84 (9): 934–40.
</div>
<div id="ref-hoelzeretal12" class="csl-entry">
Hoelzer, K., R. Pouillot, D. Gallagher, M. B. Silverman, J. Kause, and S. Dennis. 2012. <span>“<span class="nocase">Estimation of <em>Listeria Monocytogenes</em> Transfer Coefficients and Efficacy of Bacterial Removal Through Cleaning and Sanitation</span>.”</span> <em>International Journal of Food Microbiology</em> 157 (2): 267–77.
</div>
<div id="ref-Hose04" class="csl-entry">
Hose, G. C., and P. J. Van den Brink. 2004. <span>“<span class="nocase">Confirming the Species-Sensitivity Distribution Concept for Endosulfan Using Laboratory, Mesocosm, and Field Data</span>.”</span> <em><span>Archives of Environmental Contamination and Toxicology</span></em> <span>47</span> (<span>4</span>): 511–20.
</div>
<div id="ref-hyndmanfan96" class="csl-entry">
Hyndman, R. J., and Y. Fan. 1996. <span>“<span class="nocase">Sample Quantiles in Statistical Packages</span>.”</span> <em>The American Statistician</em> 50: 361–65.
</div>
<div id="ref-jaloustreetal11" class="csl-entry">
Jaloustre, S., M. Cornu, E. Morelli, V. Noel, and M. L. Delignette-Muller. 2011. <span>“<span class="nocase">Bayesian Modeling of <em>Clostridium perfringens</em> Growth in Beef-in-Sauce Products</span>.”</span> <em>Food Microbiology</em> 28 (2): 311–20.
</div>
<div id="ref-jongenburgeretal12" class="csl-entry">
Jongenburger, I., M. W. Reij, E. P. J. Boer, M. H. Zwietering, and L. G. M. Gorris. 2012. <span>“Modelling Homogeneous and Heterogeneous Microbial Contaminations in a Powdered Food Product.”</span> <em>International Journal of Food Microbiology</em> 157 (1): 35–44.
</div>
<div id="ref-jordan05" class="csl-entry">
Jordan, D. 2005. <span>“Simulating the Sensitivity of Pooled-Sample Herd Tests for Fecal Salmonella in Cattle.”</span> <em>Preventive Veterinary Medicine</em> 70 (1-2): 59–73.
</div>
<div id="ref-kefford07" class="csl-entry">
Kefford, B. J., E. J. Fields, C. Clay, and D. Nugegoda. 2007. <span>“<span class="nocase">Salinity Tolerance of Riverine Macroinvertebrates from the Southern Murray-Darling Basin</span>.”</span> <em>Marine and Freshwater Research</em> 58: 1019–31.
</div>
<div id="ref-kleinmoeschberger03" class="csl-entry">
Klein, J. P., and M. L. Moeschberger. 2003. <em>Survival Analysis: Techniques for Censored and Truncated Data</em>. 2nd ed. Springer-Verlag.
</div>
<div id="ref-Klugmanetal09" class="csl-entry">
Klugman, S. A., H. H. Panjer, and G. E. Willmot. 2009. <em>Loss Models: From Data to Decisions</em>. 3rd ed. John Wiley &amp; Sons.
</div>
<div id="ref-kochetal12" class="csl-entry">
Koch, F. H., D. Yemshanov, R. D. Magarey, and W. D. Smith. 2012. <span>“Dispersal of Invasive Forest Insects via Recreational Firewood: A Quantitative Analysis.”</span> <em>Journal of Economic Entomology</em> 105 (2): 438–50.
</div>
<div id="ref-distrModJSS" class="csl-entry">
Kohl, M., and P. Ruckdeschel. 2010. <span>“<span class="nocase">R Package : S4 Classes and Methods for Probability Models</span>.”</span> <em>Journal of Statistical Software</em> 35 (10): 1–27.
</div>
<div id="ref-larrasetal13" class="csl-entry">
Larras, Floriane, Bernard Montuelle, and Agnès Bouchez. 2013. <span>“<span class="nocase">Assessment of Toxicity Thresholds in Aquatic Environments: Does Benthic Growth of Diatoms Affect their Exposure and Sensitivity to Herbicides?</span>”</span> <em>Science of The Total Environment</em> 463-464: 469–77.
</div>
<div id="ref-lehaetal11" class="csl-entry">
Leha, A., T. Beissbarth, and K. Jung. 2011. <span>“<span class="nocase">Sequential Interim Analyses of Survival Data in <span>DNA</span> Microarray Experiments</span>.”</span> <em>BMC Bioinformatics</em> 12 (127): 1–14.
</div>
<div id="ref-luangkesornetal12" class="csl-entry">
Luangkesorn, K. L., B. A. Norman, Y. Zhuang, M. Falbo, and J. Sysko. 2012. <span>“<span class="nocase">Practice Summaries: Designing Disease Prevention and Screening Centers in Abu Dhabi</span>.”</span> <em>Interfaces</em> 42 (4): 406–9.
</div>
<div id="ref-Luceno06" class="csl-entry">
Luceno, A. 2006. <span>“<span class="nocase">Fitting the Generalized Pareto Distribution to Data Using Maximum Goodness-of-fit Estimators</span>.”</span> <em><span>Computational Statistics and Data Analysis</span></em> <span>51</span> (<span>2</span>): 904–17.
</div>
<div id="ref-mala2013publi" class="csl-entry">
Malá, I. 2013. <span>“The Use of Finite Mixtures of Lognormal and Gamma Distributions.”</span> <em>Research Journal of Economics, Business and ICT</em> 8 (2): 55–61.
</div>
<div id="ref-mandletal13" class="csl-entry">
Mandl, J. N., J. P. Monteiro, N. Vrisekoop, and R. N. Germain. 2013. <span>“<span class="nocase">T Cell-Positive Selection Uses Self-Ligand Binding Strength to Optimize Repertoire Recognition of Foreign Antigens</span>.”</span> <em>Immunity</em> <span>38</span> (<span>2</span>): 263–74.
</div>
<div id="ref-marquetouxetal12" class="csl-entry">
Marquetoux, N., M. Paul, S. Wongnarkpet, C. Poolkhet, W. Thanapongtham, F. Roger, C. Ducrot, and K. Chalvet-Monfray. 2012. <span>“<span class="nocase">Estimating Spatial and Temporal Variations of the Reproduction Number for Highly Pathogenic Avian Influenza H5N1 Epidemic in Thailand</span>.”</span> <em>Preventive Veterinary Medicine</em> 106 (2): 143–51.
</div>
<div id="ref-mcneil97" class="csl-entry">
McNeil, A. J. 1997. <span>“Estimating the Tails of Loss Severity Distributions Using Extreme Value Theory.”</span> <em>ASTIN Bulletin</em> 27 (1): 117–37.
</div>
<div id="ref-meheustetal12" class="csl-entry">
Méheust, D., P. Le Cann, T. Reponen, J. Wakefield, and S. Vesper. 2012. <span>“<span class="nocase">Possible Application of the Environmental Relative Moldiness Index in France: a Pilot Study in Brittany</span>.”</span> <em>International Journal of Hygiene and Environmental Health</em> <span>216</span> (<span>3</span>): 333–40.
</div>
<div id="ref-meyeretal13" class="csl-entry">
Meyer, W. K., S. Zhang, S. Hayakawa, H. Imai, and M. Przeworski. 2013. <span>“The Convergent Evolution of Blue Iris Pigmentation in Primates Took Distinct Molecular Paths.”</span> <em>American Journal of Physical Anthropology</em> 151 (3): 398–407.
</div>
<div id="ref-nadarajahbakar2013" class="csl-entry">
Nadarajah, S., and S. A. A. Bakar. 2013. <span>“<span class="nocase">CompLognormal: An R Package for Composite Lognormal Distributions.</span>”</span> <em>R Journal</em> 5 (2): 98–104.
</div>
<div id="ref-orellanoetal12" class="csl-entry">
Orellano, P. W., J. I. Reynoso, A. Grassi, A. Palmieri, O. Uez, and O. Carlino. 2012. <span>“<span class="nocase">Estimation of the Serial Interval for Pandemic Influenza (pH1N1) in the Most Southern Province of Argentina</span>.”</span> <em>Iranian Journal of Public Health</em> 41 (12): 26–29.
</div>
<div id="ref-Posthuma2010" class="csl-entry">
Posthuma, L., G. W. Suter, and T. P. Traas. 2010. <em>Species Sensitivity Distributions in Ecotoxicology</em>. Environmental and Ecological Risk Assessment Series. Taylor &amp; Francis.
</div>
<div id="ref-Pouillot10" class="csl-entry">
Pouillot, R., and M. L. Delignette-Muller. 2010. <span>“<span class="nocase">Evaluating Variability and Uncertainty Separately in Microbial Quantitative Risk Assessment using two R Packages</span>.”</span> <em><span>International Journal of Food Microbiology</span></em> <span>142</span> (<span>3</span>): 330–40.
</div>
<div id="ref-mc2d" class="csl-entry">
Pouillot, R., M. L. Delignette-Muller, and J. B. Denis. 2011. <em>: Tools for Two-Dimensional Monte-Carlo Simulations</em>. <a href="https://cran.r-project.org/package=mc2d" class="external-link">https://cran.r-project.org/package=mc2d</a>.
</div>
<div id="ref-poulliotetal12" class="csl-entry">
Pouillot, R., K. Hoelzer, Y. Chen, and S. Dennis. 2012. <span>“<span class="nocase">Estimating Probability Distributions of Bacterial Concentrations in Food Based on Data Generated Using the Most Probable Number (MPN) Method for Use in Risk Assessment</span>.”</span> <em>Food Control</em> 29 (2): 350–57.
</div>
<div id="ref-prosseretal13" class="csl-entry">
Prosser, D. J., L. L. Hungerford, R. M. Erwin, M. A. Ottinger, J. Y. Takekawa, and E. C. Ellis. 2013. <span>“Mapping Avian Influenza Transmission Risk at the Interface of Domestic Poultry and Wild Birds.”</span> <em>Frontiers in Public Health</em> 1 (28): 1–11.
</div>
<div id="ref-R13" class="csl-entry">
R Development Core Team. 2013. <em>R: A Language and Environment for Statistical Computing</em>. Vienna, Austria. <a href="https://www.r-project.org/" class="external-link">https://www.r-project.org/</a>.
</div>
<div id="ref-Ricci05" class="csl-entry">
Ricci, V. 2005. <span>“Fitting Distributions with r.”</span> <a href="https://cran.r-project.org/doc/contrib/Ricci-distributions-en.pdf" class="external-link">https://cran.r-project.org/doc/contrib/Ricci-distributions-en.pdf</a>.
</div>
<div id="ref-Rigaux2014" class="csl-entry">
Rigaux, Clémence, Stéphane André, Isabelle Albert, and Frédéric Carlin. 2014. <span>“Quantitative Assessment of the Risk of Microbial Spoilage in Foods. Prediction of Non-Stability at 55<span class="math inline">\(\,^{\circ}\)</span>c Caused by <em>Geobacillus Stearothermophilus</em> in Canned Green Beans.”</span> <em>International Journal of Food Microbiology</em> 171: 119–28.
</div>
<div id="ref-saketal11" class="csl-entry">
Sak, H., and C. Haksoz. 2011. <span>“A Copula-Based Simulation Model for Supply Portfolio Risk.”</span> <em>Journal of Operational Risk</em> 6 (3): 15–38.
</div>
<div id="ref-sosaetal2013" class="csl-entry">
Samuel-Rosa, A., R. Simao Diniz Dalmolin, and P. Miguel. 2013. <span>“Building Predictive Models of Soil Particle-Size Distribution.”</span> <em>Revista Brasileira de Ciencia Do Solo</em> 37: 422–30.
</div>
<div id="ref-satoetal13" class="csl-entry">
Sato, Maria Ines Z., Ana Tereza Galvani, Jose Antonio Padula, Adelaide Cassia Nardocci, Marcelo de Souza Lauretto, Maria Tereza Pepe Razzolini, and Elayse Maria Hachich. 2013. <span>“<span class="nocase">Assessing the Infection Risk of <em>Giardia</em> and <em>Cryptosporidium</em> in Public Drinking Water Delivered by Surface Water Systems in Sao Paulo State, Brazil</span>.”</span> <em>Science of The Total Environment</em> 442: 389–96.
</div>
<div id="ref-scholletal12" class="csl-entry">
Scholl, C. F., C. C. Nice, J. A. Fordyce, Z. Gompert, and M. L. Forister. 2012. <span>“Larval Performance in the Context of Ecological Diversification and Speciation in Lycaeides Butterflies.”</span> <em>International Journal of Ecology</em> 2012 (ID 242154): 1–13.
</div>
<div id="ref-simoetal13" class="csl-entry">
Simó, J., Francesc Casaña, and J. Sabaté. 2013. <span>“<span class="nocase">Modelling <span>‘calçots’</span> (<em>Alium cepa L.</em>) Growth by Gompertz Function</span>.”</span> <em>Statistics and Operations Research Transactions</em> 37 (1): 95–106.
</div>
<div id="ref-srinivasanetal2013" class="csl-entry">
Srinivasan, S., T. P. Sorrell, J. P. Brooks, D. J. Edwards, and R. Diehl McDougle. 2013. <span>“<span class="nocase">Workforce Assessment Method for an Urban Police Department: Using Analytics to Estimate Patrol Staffing</span>.”</span> <em>Policing: An International Journal of Police Strategies &amp; Management</em> 36 (4): 702–18.
</div>
<div id="ref-staggeetal13" class="csl-entry">
Stagge, J. H., and G. E. Moglen. 2013. <span>“A Nonparametric Stochastic Method for Generating Daily Climate-Adjusted Streamflows.”</span> <em>Water Resources Research</em> 49 (10): 6179–93.
</div>
<div id="ref-suuronenetal12" class="csl-entry">
Suuronen, J. P., A. Kallonen, M. Eik, J. Puttonen, Ritva Serimaa, and Heiko Herrmann. 2012. <span>“Analysis of Short Fibres Orientation in Steel Fibre-Reinforced Concrete (SFRC) by x-Ray Tomography.”</span> <em>Journal of Materials Science</em> 48 (3): 1358–67.
</div>
<div id="ref-tarnczi11" class="csl-entry">
Tarnczi, T., V. Fenyves, and Z. Bcs. 2011. <span>“The Business Uncertainty and Variability Management with Real Options Models Combined Two Dimensional Simulation.”</span> <em>International Journal of Management Cases</em> 13 (3): 159–67.
</div>
<div id="ref-telloetal12" class="csl-entry">
Tello, A., B. Austin, and T. C. Telfer. 2012. <span>“Selective Pressure of Antibiotic Pollution on Bacteria of Importance to Public Health.”</span> <em>Environmental Health Perspectives</em> 120 (8): 1100–1106.
</div>
<div id="ref-survival" class="csl-entry">
Therneau, T. 2011. <em>: Survival Analysis, Including Penalized Likelihood</em>. <a href="https://cran.r-project.org/package=survival" class="external-link">https://cran.r-project.org/package=survival</a>.
</div>
<div id="ref-tikoleetal13" class="csl-entry">
Tikole, S., V. Jaravine, V. Yu Orekhov, and P. Guentert. 2013. <span>“<span class="nocase">Effects of NMR spectral resolution on protein structure calculation</span>.”</span> <em>PloS One</em> 8 (7): e68567.
</div>
<div id="ref-Tse2009" class="csl-entry">
Tse, Y. K. 2009. <em>Nonlife Actuarial Models: Theory, Methods and Evaluation</em>. 1st ed. International Series on Actuarial Science. Cambridge University Press.
</div>
<div id="ref-Turnbull74" class="csl-entry">
Turnbull, B. W. 1974. <span>“<span class="nocase">Nonparametric Estimation of a Survivorship Function with Doubly Censored Data</span>.”</span> <em><span>Journal of the American Statistical Association</span></em> <span>69</span> (<span>345</span>): 169–73.
</div>
<div id="ref-vaninsky13" class="csl-entry">
Vaninsky, A. Y. 2013. <span>“<span class="nocase">Stochastic DEA with a Perfect Object and Its Application to Analysis of Environmental Efficiency</span>.”</span> <em>American Journal of Applied Mathematics and Statistics</em> 1 (4): 57–63.
</div>
<div id="ref-MASS" class="csl-entry">
Venables, W. N., and B. D. Ripley. 2010. <em>Modern Applied Statistics with <span>S</span></em>. 4th ed. Springer-Verlag.
</div>
<div id="ref-vianaetal13" class="csl-entry">
Viana, D. S., L. Santamará, T. C. Michot, and J. Figuerola. 2013. <span>“Allometric Scaling of Long-Distance Seed Dispersal by Migratory Birds.”</span> <em>The American Naturalist</em> 181 (5): 649–62.
</div>
<div id="ref-voigtetal14" class="csl-entry">
Voigt, Christian C., Linn S. Lehnert, Ana G. Popa-Lisseanu, Mateusz Ciechanowski, Péter Estók, Florian Gloza-Rausch, Tamás Goerfoel, et al. 2014. <span>“<span class="nocase">The Trans-Boundary Importance of Artificial Bat <em>hibernacula</em> in Managed European Forests</span>.”</span> <em>Biodiversity and Conservation</em> 23: 617–31.
</div>
<div id="ref-Vose10" class="csl-entry">
Vose, D. 2010. <em>Quantitative Risk Analysis. A Guide to Monte Carlo Simulation Modelling</em>. 1st ed. John Wiley &amp; Sons.
</div>
<div id="ref-Wang2007" class="csl-entry">
Wang, Yong. 2007. <span>“On Fast Computation of the Non-Parametric Maximum Likelihood Estimate of a Mixing Distribution.”</span> <em>Journal of the Royal Statistical Society: Series B (Statistical Methodology)</em> 69 (2): 185–98.
</div>
<div id="ref-Wang2008" class="csl-entry">
———. 2008. <span>“Dimension-Reduced Nonparametric Maximum Likelihood Computation for Interval-Censored Data.”</span> <em>Computational Statistics &amp; Data Analysis</em> 52 (5): 2388–2402.
</div>
<div id="ref-Wang2018" class="csl-entry">
Wang, Yong, and Shabnam Fani. 2018. <span>“Nonparametric Maximum Likelihood Computation of a u-Shaped Hazard Function.”</span> <em>Statistics and Computing</em> 28 (1): 187–200.
</div>
<div id="ref-Wang2013" class="csl-entry">
Wang, Yong, and Stephen M Taylor. 2013. <span>“Efficient Computation of Nonparametric Survival Functions via a Hierarchical Mixture Formulation.”</span> <em>Statistics and Computing</em> 23 (6): 713–25.
</div>
<div id="ref-wayland13" class="csl-entry">
Wayland, M. T. 2013. <span>“<span class="nocase">Morphological Variation in <em>Echinorhynchus truttae</em> Schrank, 1788 and the <em>Echinorhynchus bothniensis</em> Zdzitowiecki &amp; Valtonen, 1987 species complex from freshwater fishes of northern Europe</span>.”</span> <em>Biodiversity Data Journal</em> 1: e975.
</div>
<div id="ref-westphalfitch13" class="csl-entry">
Westphal-Fitch, G., and W. T. Fitch. 2013. <span>“Spatial Analysis of <span>‘Crazy Quilts,’</span> a Class of Potentially Random Aesthetic Artefacts.”</span> <em>PloS One</em> 8 (9): e74055.
</div>
<div id="ref-wu13" class="csl-entry">
Wu, Xing Zheng. 2013a. <span>“Probabilistic Slope Stability Analysis by a Copula-Based Sampling Method.”</span> <em>Computational Geosciences</em> 17 (5): 739–55.
</div>
<div id="ref-Wu2013b" class="csl-entry">
———. 2013b. <span>“Trivariate Analysis of Soil Ranking-Correlated Characteristics and Its Application to Probabilistic Stability Assessments in Geotechnical Engineering Problems.”</span> <em>Soils and Foundations</em> 53 (4): 540–56.
</div>
<div id="ref-Zhang2013" class="csl-entry">
Zhang, Yu, Emad Habib, Robert J. Kuligowski, and Dongsoo Kim. 2013. <span>“Joint Distribution of Multiplicative Errors in Radar and Satellite {QPEs} and Its Use in Estimating the Conditional Exceedance Probability.”</span> <em>Advances in Water Resources</em> 59: 133–45.
</div>
</div>
</div>

  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Marie-Laure Delignette-Muller, Christophe Dutang, Aurelie Siberchicot.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
